--- a/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/net/minecraft/world/entity/animal/camel/Camel.java
@@ -48,7 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.Vec2;
-import net.minecraft.world.phys.Vec3;
+import net.minecraft.world.phys.Vec3;import org.bukkit.event.entity.EntityDamageEvent;
 
 public class Camel extends AbstractHorse implements PlayerRideableJumping, Saddleable {
     public static final float BABY_SCALE = 0.45F;
@@ -83,6 +_,17 @@
         groundpathnavigation.setCanWalkOverFences(true);
     }
 
+    // Purpur start
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.camelRidableInWater;
+    }
+
+    public int getPurpurBreedTime() {
+        return this.level().purpurConfig.camelBreedingTicks;
+    }
+    // Purpur end
+
     @Override
     public void addAdditionalSaveData(CompoundTag p_250330_) {
         super.addAdditionalSaveData(p_250330_);
@@ -286,6 +_,7 @@
         this.dashCooldown = 55;
         this.setDashing(true);
         this.hasImpulse = true;
+        net.neoforged.neoforge.common.CommonHooks.onLivingJump(this);
     }
 
     public boolean isDashing() {
@@ -318,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.CAMEL_DEATH;
     }
 
@@ -451,9 +_,14 @@
     }
 
     @Override
-    protected void actuallyHurt(DamageSource p_250410_, float p_251451_) {
+    public boolean actuallyHurt(DamageSource source, float amount, EntityDamageEvent event) { // CraftBukkit start - void -> boolean
+        final var result = super.actuallyHurt(source, amount, event); // CraftBukkit
+        if (!result) {
+            return false;
+        }
+        // CraftBukkit end
         this.standUpInstantly();
-        super.actuallyHurt(p_250410_, p_251451_);
+        return true; // CraftBukkit
     }
 
     @Override
@@ -556,7 +_,7 @@
     }
 
     public void sitDown() {
-        if (!this.isCamelSitting()) {
+        if (!this.isCamelSitting() && new io.papermc.paper.event.entity.EntityToggleSitEvent(this.getBukkitEntity(), true).callEvent()) { // Paper - Add EntityToggleSitEvent
             this.makeSound(SoundEvents.CAMEL_SIT);
             this.setPose(Pose.SITTING);
             this.gameEvent(GameEvent.ENTITY_ACTION);
@@ -565,7 +_,7 @@
     }
 
     public void standUp() {
-        if (this.isCamelSitting()) {
+        if (this.isCamelSitting() && new io.papermc.paper.event.entity.EntityToggleSitEvent(this.getBukkitEntity(), false).callEvent()) { // Paper - Add EntityToggleSitEvent
             this.makeSound(SoundEvents.CAMEL_STAND);
             this.setPose(Pose.STANDING);
             this.gameEvent(GameEvent.ENTITY_ACTION);
@@ -574,6 +_,7 @@
     }
 
     public void standUpInstantly() {
+        if (this.isCamelSitting() && !new io.papermc.paper.event.entity.EntityToggleSitEvent(this.getBukkitEntity(), false).callEvent()) return; // Paper - Add EntityToggleSitEvent
         this.setPose(Pose.STANDING);
         this.gameEvent(GameEvent.ENTITY_ACTION);
         this.resetLastPoseChangeTickToFullStand(this.level().getGameTime());
