--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -190,7 +_,7 @@
     }
 
     @Override
-    protected float getSoundVolume() {
+    public float getSoundVolume() {
         return 4.0F;
     }
 
@@ -206,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.WARDEN_DEATH;
     }
 
@@ -407,7 +_,7 @@
 
     public static void applyDarknessAround(ServerLevel p_219376_, Vec3 p_219377_, @Nullable Entity p_219378_, int p_219379_) {
         MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.DARKNESS, 260, 0, false, false);
-        MobEffectUtil.addEffectToPlayersAround(p_219376_, p_219378_, p_219377_, (double)p_219379_, mobeffectinstance, 200);
+        MobEffectUtil.addEffectToPlayersAround(p_219376_, p_219378_, p_219377_, (double)p_219379_, mobeffectinstance, 200, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.WARDEN); // CraftBukkit - Add EntityPotionEffectEvent.Cause
     }
 
     @Override
@@ -469,6 +_,15 @@
     @VisibleForTesting
     public void increaseAngerAt(@Nullable Entity p_219388_, int p_219389_, boolean p_219390_) {
         if (!this.isNoAi() && this.canTargetEntity(p_219388_)) {
+            // Paper start - Add WardenAngerChangeEvent
+            int activeAnger = this.angerManagement.getActiveAnger(p_219388_);
+            io.papermc.paper.event.entity.WardenAngerChangeEvent event = new io.papermc.paper.event.entity.WardenAngerChangeEvent((org.bukkit.entity.Warden) this.getBukkitEntity(), p_219388_.getBukkitEntity(), activeAnger, Math.min(150, activeAnger + p_219389_));
+            this.level().getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
+            p_219389_ = event.getNewAnger() - activeAnger;
+            // Paper end - Add WardenAngerChangeEvent
             WardenAi.setDigCooldown(this);
             boolean flag = !(this.getTarget() instanceof Player);
             int i = this.angerManagement.increaseAnger(p_219388_, p_219389_);
@@ -511,14 +_,14 @@
     }
 
     @Override
-    public boolean hurt(DamageSource p_219381_, float p_219382_) {
-        boolean flag = super.hurt(p_219381_, p_219382_);
+    public boolean hurt(DamageSource p_21016_, float p_21017_) {
+        boolean flag = super.hurt(p_21016_, p_21017_);
         if (!this.level().isClientSide && !this.isNoAi() && !this.isDiggingOrEmerging()) {
-            Entity entity = p_219381_.getEntity();
+            Entity entity = p_21016_.getEntity();
             this.increaseAngerAt(entity, AngerLevel.ANGRY.getMinimumAnger() + 20, false);
             if (this.brain.getMemory(MemoryModuleType.ATTACK_TARGET).isEmpty()
                 && entity instanceof LivingEntity livingentity
-                && (p_219381_.isDirect() || this.closerThan(livingentity, 5.0))) {
+                && (p_21016_.isDirect() || this.closerThan(livingentity, 5.0))) {
                 this.setAttackTarget(livingentity);
             }
         }
