--- a/net/minecraft/world/level/EntityGetter.java
+++ b/net/minecraft/world/level/EntityGetter.java
@@ -18,7 +_,7 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public interface EntityGetter {
+public interface EntityGetter extends ca.spottedleaf.moonrise.patches.chunk_system.world.ChunkSystemEntityGetter { // Paper - rewrite chunk system
     List<Entity> getEntities(@Nullable Entity p_45936_, AABB p_45937_, Predicate<? super Entity> p_45938_);
 
     <T extends Entity> List<T> getEntities(EntityTypeTest<Entity, T> p_151464_, AABB p_151465_, Predicate<? super T> p_151466_);
@@ -33,21 +_,44 @@
         return this.getEntities(p_45934_, p_45935_, EntitySelector.NO_SPECTATORS);
     }
 
-    default boolean isUnobstructed(@Nullable Entity p_45939_, VoxelShape p_45940_) {
-        if (p_45940_.isEmpty()) {
-            return true;
-        } else {
-            for (Entity entity : this.getEntities(p_45939_, p_45940_.bounds())) {
-                if (!entity.isRemoved()
-                    && entity.blocksBuilding
-                    && (p_45939_ == null || !entity.isPassengerOfSameVehicle(p_45939_))
-                    && Shapes.joinIsNotEmpty(p_45940_, Shapes.create(entity.getBoundingBox()), BooleanOp.AND)) {
-                    return false;
+    // Paper start - rewrite chunk system
+    @Override
+    default List<Entity> moonrise$getHardCollidingEntities(final Entity entity, final AABB box, final Predicate<? super Entity> predicate) {
+        return this.getEntities(entity, box, predicate);
+    }
+    // Paper end - rewrite chunk system
+
+    // Paper start - optimise collisions
+    default boolean isUnobstructed(@Nullable Entity entity, VoxelShape voxel) {
+        if (voxel.isEmpty()) {
+            return false;
+        }
+
+        final AABB singleAABB = ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)voxel).moonrise$getSingleAABBRepresentation();
+        final List<Entity> entities = this.getEntities(
+            entity,
+            singleAABB == null ? voxel.bounds() : singleAABB.inflate(-ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_EPSILON, -ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_EPSILON, -ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_EPSILON)
+        );
+
+        for (int i = 0, len = entities.size(); i < len; ++i) {
+            final Entity otherEntity = entities.get(i);
+
+            if (otherEntity.isRemoved() || !otherEntity.blocksBuilding || (entity != null && otherEntity.isPassengerOfSameVehicle(entity))) {
+                continue;
+            }
+
+            if (singleAABB == null) {
+                final AABB entityBB = otherEntity.getBoundingBox();
+                if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isEmpty(entityBB) || !ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersectNoEmpty(voxel, entityBB)) {
+                    continue;
                 }
             }
 
-            return true;
+            return false;
         }
+
+        return true;
+        // Paper end - optimise collisions
     }
 
     default <T extends Entity> List<T> getEntitiesOfClass(Class<T> p_45977_, AABB p_45978_) {
@@ -55,57 +_,135 @@
     }
 
     default List<VoxelShape> getEntityCollisions(@Nullable Entity p_186451_, AABB p_186452_) {
-        if (p_186452_.getSize() < 1.0E-7) {
-            return List.of();
-        } else {
-            Predicate<Entity> predicate = p_186451_ == null ? EntitySelector.CAN_BE_COLLIDED_WITH : EntitySelector.NO_SPECTATORS.and(p_186451_::canCollideWith);
-            List<Entity> list = this.getEntities(p_186451_, p_186452_.inflate(1.0E-7), predicate);
-            if (list.isEmpty()) {
-                return List.of();
-            } else {
-                Builder<VoxelShape> builder = ImmutableList.builderWithExpectedSize(list.size());
-
-                for (Entity entity : list) {
-                    builder.add(Shapes.create(entity.getBoundingBox()));
-                }
-
-                return builder.build();
-            }
-        }
+        // Paper start - optimise collisions
+      // first behavior change is to correctly check for empty AABB
+      if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isEmpty(p_186452_)) {
+          // reduce indirection by always returning type with same class
+          return new java.util.ArrayList<>();
+      }
+
+      // to comply with vanilla intersection rules, expand by -epsilon so that we only get stuff we definitely collide with.
+      // Vanilla for hard collisions has this backwards, and they expand by +epsilon but this causes terrible problems
+      // specifically with boat collisions.
+      p_186452_ = p_186452_.inflate(-ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_EPSILON, -ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_EPSILON, -ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_EPSILON);
+
+      final List<Entity> entities;
+      if (p_186451_ != null && ((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)p_186451_).moonrise$isHardColliding()) {
+          entities = this.getEntities(p_186451_, p_186452_, null);
+      } else {
+          entities = ((ca.spottedleaf.moonrise.patches.chunk_system.world.ChunkSystemEntityGetter)this).moonrise$getHardCollidingEntities(p_186451_, p_186452_, null);
+      }
+
+      final List<VoxelShape> ret = new java.util.ArrayList<>(Math.min(25, entities.size()));
+
+      for (final Entity otherEntity : entities) { // Stellar - enhance for-each loop
+          if (otherEntity.isSpectator()) {
+              continue;
+          }
+
+          if ((p_186451_ == null && otherEntity.canBeCollidedWith()) || (p_186451_ != null && p_186451_.canCollideWith(otherEntity))) {
+              ret.add(Shapes.create(otherEntity.getBoundingBox()));
+          }
+      }
+
+      return ret;
+      // Paper end - optimise collisions
     }
 
     @Nullable
     default Player getNearestPlayer(double p_45919_, double p_45920_, double p_45921_, double p_45922_, @Nullable Predicate<Entity> p_45923_) {
-        double d0 = -1.0;
+        // Stellar start - Leaf: Slightly optimise getNearestPlayer
+        if (p_45922_ < 0.0) {
+            p_45922_ = Double.MAX_VALUE;
+        } else {
+            p_45922_ = p_45922_ * p_45922_;
+        }
+        // Stellar end - Leaf: Slightly optimise getNearestPlayer
         Player player = null;
 
-        for (Player player1 : this.players()) {
-            if (p_45923_ == null || p_45923_.test(player1)) {
+        // Stellar start - Leaf: Slightly optimise getNearestPlayer
+        List<? extends Player> players = this.players();
+        if (p_45923_ == null) {
+        for (final Player player1 : players) {
+        // Stellar end - Leaf: Slightly optimise getNearestPlayer
                 double d1 = player1.distanceToSqr(p_45919_, p_45920_, p_45921_);
-                if ((p_45922_ < 0.0 || d1 < p_45922_ * p_45922_) && (d0 == -1.0 || d1 < d0)) {
-                    d0 = d1;
+                if (d1 < p_45922_) { // Stellar - Leaf: Slightly optimise getNearestPlayer
+                    p_45922_ = d1; // Stellar - Leaf: Slightly optimise getNearestPlayer
                     player = player1;
                 }
+        // Stellar start - Leaf: Slightly optimise getNearestPlayer
+        }} else {
+            for (final Player player1 : players) {
+                if (p_45923_.test(player1)) {
+                    double d1 = player1.distanceToSqr(p_45919_, p_45920_, p_45921_);
+                    if (d1 < p_45922_) {
+                        p_45922_ = d1;
+                        player = player1;
+                    }
+                }
+        // Stellar end - Leaf: Slightly optimise getNearestPlayer
             }
         }
 
         return player;
     }
 
+    // Paper start
+    default List<org.bukkit.entity.HumanEntity> findNearbyBukkitPlayers(double x, double y, double z, double radius, boolean notSpectator) {
+        return findNearbyBukkitPlayers(x, y, z, radius, notSpectator ? EntitySelector.NO_SPECTATORS : net.minecraft.world.entity.EntitySelector.NO_CREATIVE_OR_SPECTATOR);
+    }
+
+    default List<org.bukkit.entity.HumanEntity> findNearbyBukkitPlayers(double x, double y, double z, double radius, @Nullable Predicate<Entity> predicate) {
+        com.google.common.collect.ImmutableList.Builder<org.bukkit.entity.HumanEntity> builder = com.google.common.collect.ImmutableList.builder();
+
+        for (Player human : this.players()) {
+            if (predicate == null || predicate.test(human)) {
+                double distanceSquared = human.distanceToSqr(x, y, z);
+
+                if (radius < 0.0D || distanceSquared < radius * radius) {
+                    builder.add(human.getBukkitEntity());
+                }
+            }
+        }
+
+        return builder.build();
+    }
+    // Paper end
+
     @Nullable
     default Player getNearestPlayer(Entity p_45931_, double p_45932_) {
         return this.getNearestPlayer(p_45931_.getX(), p_45931_.getY(), p_45931_.getZ(), p_45932_, false);
     }
 
+    // Paper start - Affects Spawning API
+    default @Nullable Player findNearbyPlayer(Entity entity, double maxDistance, @Nullable Predicate<Entity> predicate) {
+        return this.getNearestPlayer(entity.getX(), entity.getY(), entity.getZ(), maxDistance, predicate);
+    }
+    // Paper end - Affects Spawning API
+
     @Nullable
     default Player getNearestPlayer(double p_45925_, double p_45926_, double p_45927_, double p_45928_, boolean p_45929_) {
         Predicate<Entity> predicate = p_45929_ ? EntitySelector.NO_CREATIVE_OR_SPECTATOR : EntitySelector.NO_SPECTATORS;
         return this.getNearestPlayer(p_45925_, p_45926_, p_45927_, p_45928_, predicate);
     }
 
+    // Paper start - Affects Spawning API
+    default boolean hasNearbyAlivePlayerThatAffectsSpawning(double x, double y, double z, double range) {
+        for (Player player : this.players()) {
+            if (EntitySelector.PLAYER_AFFECTS_SPAWNING.test(player)) { // combines NO_SPECTATORS and LIVING_ENTITY_STILL_ALIVE with an "affects spawning" check
+                double distanceSqr = player.distanceToSqr(x, y, z);
+                if (range < 0.0D || distanceSqr < range * range) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    // Paper end - Affects Spawning API
+
     default boolean hasNearbyAlivePlayer(double p_45915_, double p_45916_, double p_45917_, double p_45918_) {
         for (Player player : this.players()) {
-            if (EntitySelector.NO_SPECTATORS.test(player) && EntitySelector.LIVING_ENTITY_STILL_ALIVE.test(player)) {
+            if (EntitySelector.NO_SPECTATORS.test(player) && EntitySelector.LIVING_ENTITY_STILL_ALIVE.test(player) && EntitySelector.notAfk.test(player)) { // Purpur
                 double d0 = player.distanceToSqr(p_45915_, p_45916_, p_45917_);
                 if (p_45918_ < 0.0 || d0 < p_45918_ * p_45918_) {
                     return true;
