--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -124,6 +_,15 @@
             if (this.canSpreadTo(
                 p_255851_, p_76012_, blockstate, Direction.DOWN, blockpos, blockstate1, p_255851_.getFluidState(blockpos), fluidstate.getType()
             )) {
+                // CraftBukkit start
+                org.bukkit.block.Block source = org.bukkit.craftbukkit.block.CraftBlock.at(p_255851_, p_76012_);
+                org.bukkit.event.block.BlockFromToEvent event = new org.bukkit.event.block.BlockFromToEvent(source, org.bukkit.block.BlockFace.DOWN);
+                p_255851_.getCraftServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+                // CraftBukkit end
                 this.spreadTo(p_255851_, blockpos, blockstate1, Direction.DOWN, fluidstate);
                 if (this.sourceNeighborCount(p_255851_, p_76012_) >= 3) {
                     this.spreadToSides(p_255851_, p_76012_, p_76013_, blockstate);
@@ -147,8 +_,18 @@
                 Direction direction = entry.getKey();
                 FluidState fluidstate = entry.getValue();
                 BlockPos blockpos = p_76016_.relative(direction);
-                BlockState blockstate = p_256644_.getBlockState(blockpos);
+                BlockState blockstate = p_256644_.getBlockStateIfLoaded(blockpos); // Paper - Prevent chunk loading from fluid flowing
+                if (blockstate == null) continue; // Paper - Prevent chunk loading from fluid flowing
                 if (this.canSpreadTo(p_256644_, p_76016_, p_76018_, direction, blockpos, blockstate, p_256644_.getFluidState(blockpos), fluidstate.getType())) {
+                    // CraftBukkit start
+                    org.bukkit.block.Block source = org.bukkit.craftbukkit.block.CraftBlock.at(p_256644_, p_76016_);
+                    org.bukkit.event.block.BlockFromToEvent event = new org.bukkit.event.block.BlockFromToEvent(source, org.bukkit.craftbukkit.block.CraftBlock.notchToBlockFace(direction));
+                    p_256644_.getCraftServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        continue;
+                    }
+                    // CraftBukkit end
                     this.spreadTo(p_256644_, blockpos, blockstate, direction, fluidstate);
                 }
             }
@@ -161,10 +_,11 @@
 
         for (Direction direction : Direction.Plane.HORIZONTAL) {
             BlockPos blockpos = p_76037_.relative(direction);
-            BlockState blockstate = p_256464_.getBlockState(blockpos);
+            BlockState blockstate = p_256464_.getBlockStateIfLoaded(blockpos); // Paper - Prevent chunk loading from fluid flowing
+            if (blockstate == null) continue; // Paper - Prevent chunk loading from fluid flowing
             FluidState fluidstate = blockstate.getFluidState();
             if (fluidstate.getType().isSame(this) && this.canPassThroughWall(direction, p_256464_, p_76037_, p_76038_, blockpos, blockstate)) {
-                if (fluidstate.isSource()) {
+                if (fluidstate.isSource() && net.neoforged.neoforge.event.EventHooks.canCreateFluidSource(p_256464_, blockpos, blockstate)) {
                     j++;
                 }
 
@@ -172,7 +_,7 @@
             }
         }
 
-        if (this.canConvertToSource(p_256464_) && j >= 2) {
+        if (j >= 2) {
             BlockState blockstate1 = p_256464_.getBlockState(p_76037_.below());
             FluidState fluidstate1 = blockstate1.getFluidState();
             if (blockstate1.isSolid() || this.isSourceBlockOfThisType(fluidstate1)) {
@@ -238,6 +_,15 @@
         return this.getSource().defaultFluidState().setValue(FALLING, Boolean.valueOf(p_76069_));
     }
 
+    @Override
+    public boolean canConvertToSource(FluidState state, Level level, BlockPos pos) {
+        return this.canConvertToSource(level);
+    }
+
+    /**
+     * @deprecated Forge: Use {@link #canConvertToSource(FluidState, Level, BlockPos)} instead.
+     */
+    @Deprecated
     protected abstract boolean canConvertToSource(Level p_256009_);
 
     protected void spreadTo(LevelAccessor p_76005_, BlockPos p_76006_, BlockState p_76007_, Direction p_76008_, FluidState p_76009_) {
@@ -245,13 +_,14 @@
             ((LiquidBlockContainer)p_76007_.getBlock()).placeLiquid(p_76005_, p_76006_, p_76007_, p_76009_);
         } else {
             if (!p_76007_.isAir()) {
-                this.beforeDestroyingBlock(p_76005_, p_76006_, p_76007_);
+                this.beforeDestroyingBlock(p_76005_, p_76006_, p_76007_, p_76006_.relative(p_76008_.getOpposite())); // Paper - Add BlockBreakBlockEvent
             }
 
             p_76005_.setBlock(p_76006_, p_76009_.createLegacyBlock(), 3);
         }
     }
 
+    protected void beforeDestroyingBlock(LevelAccessor world, BlockPos pos, BlockState state, BlockPos source) { beforeDestroyingBlock(world, pos, state); } // Paper - Add BlockBreakBlockEvent
     protected abstract void beforeDestroyingBlock(LevelAccessor p_76002_, BlockPos p_76003_, BlockState p_76004_);
 
     private static short getCacheKey(BlockPos p_76059_, BlockPos p_76060_) {
@@ -276,10 +_,18 @@
             if (direction != p_76030_) {
                 BlockPos blockpos = p_76028_.relative(direction);
                 short short1 = getCacheKey(p_76032_, blockpos);
-                Pair<BlockState, FluidState> pair = p_76033_.computeIfAbsent(short1, p_284932_ -> {
-                    BlockState blockstate1 = p_76027_.getBlockState(blockpos);
-                    return Pair.of(blockstate1, blockstate1.getFluidState());
-                });
+                // Paper start - Prevent chunk loading from fluid flowing
+                Pair<BlockState, FluidState> pair = p_76033_.get(short1);
+                if (pair == null) {
+                    BlockState iblockdatax = p_76027_.getBlockStateIfLoaded(blockpos);
+                    if (iblockdatax == null) {
+                        continue;
+                    }
+
+                    pair = Pair.of(iblockdatax, iblockdatax.getFluidState());
+                    p_76033_.put(short1, pair);
+                }
+                // Paper end - Prevent chunk loading from fluid flowing
                 BlockState blockstate = pair.getFirst();
                 FluidState fluidstate = pair.getSecond();
                 if (this.canPassThrough(p_76027_, this.getFlowing(), p_76028_, p_76031_, direction, blockpos, blockstate, fluidstate)) {
@@ -357,10 +_,16 @@
         for (Direction direction : Direction.Plane.HORIZONTAL) {
             BlockPos blockpos = p_76081_.relative(direction);
             short short1 = getCacheKey(p_76081_, blockpos);
-            Pair<BlockState, FluidState> pair = short2objectmap.computeIfAbsent(short1, p_284929_ -> {
-                BlockState blockstate1 = p_256191_.getBlockState(blockpos);
-                return Pair.of(blockstate1, blockstate1.getFluidState());
-            });
+            // Paper start - Prevent chunk loading from fluid flowing
+            Pair<BlockState, FluidState> pair = (Pair) short2objectmap.get(short1);
+            if (pair == null) {
+                BlockState iblockdatax = p_256191_.getBlockStateIfLoaded(blockpos);
+                if (iblockdatax == null) continue;
+
+                pair = Pair.of(iblockdatax, iblockdatax.getFluidState());
+                short2objectmap.put(short1, pair);
+            }
+            // Paper end - Prevent chunk loading from fluid flowing
             BlockState blockstate = pair.getFirst();
             FluidState fluidstate = pair.getSecond();
             FluidState fluidstate1 = this.getNewLiquid(p_256191_, blockpos, blockstate);
@@ -439,11 +_,23 @@
             int i = this.getSpreadDelay(p_75995_, p_75996_, p_75997_, fluidstate);
             if (fluidstate.isEmpty()) {
                 p_75997_ = fluidstate;
-                p_75995_.setBlock(p_75996_, Blocks.AIR.defaultBlockState(), 3);
+                // CraftBukkit start
+                org.bukkit.event.block.FluidLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFluidLevelChangeEvent(p_75995_, p_75996_, Blocks.AIR.defaultBlockState());
+                if (event.isCancelled()) {
+                    return;
+                }
+                p_75995_.setBlock(p_75996_, ((org.bukkit.craftbukkit.block.data.CraftBlockData) event.getNewData()).getState(), 3);
+                // CraftBukkit end
             } else if (!fluidstate.equals(p_75997_)) {
                 p_75997_ = fluidstate;
                 BlockState blockstate = fluidstate.createLegacyBlock();
-                p_75995_.setBlock(p_75996_, blockstate, 2);
+                // CraftBukkit start
+                org.bukkit.event.block.FluidLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFluidLevelChangeEvent(p_75995_, p_75996_, blockstate);
+                if (event.isCancelled()) {
+                    return;
+                }
+                p_75995_.setBlock(p_75996_, ((org.bukkit.craftbukkit.block.data.CraftBlockData) event.getNewData()).getState(), 2);
+                // CraftBukkit end
                 p_75995_.scheduleTick(p_75996_, fluidstate.getType(), i);
                 p_75995_.updateNeighborsAt(p_75996_, blockstate.getBlock());
             }
