--- a/net/minecraft/world/entity/GlowSquid.java
+++ b/net/minecraft/world/entity/GlowSquid.java
@@ -23,6 +_,19 @@
         super(p_147111_, p_147112_);
     }
 
+    // Purpur start
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.glowSquidRidable;
+    }
+
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.glowSquidControllable;
+    }
+    // Purpur end
+
     @Override
     protected ParticleOptions getInkParticle() {
         return ParticleTypes.GLOW_SQUID_INK;
@@ -50,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.GLOW_SQUID_DEATH;
     }
 
@@ -78,8 +_,8 @@
     }
 
     @Override
-    public boolean hurt(DamageSource p_147114_, float p_147115_) {
-        boolean flag = super.hurt(p_147114_, p_147115_);
+    public boolean hurt(DamageSource source, float amount) {
+        boolean flag = super.hurt(source, amount);
         if (flag) {
             this.setDarkTicks(100);
         }
