--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -62,10 +_,10 @@
     private final BlockPos spawnPos;
     private final Stopwatch stopwatch = Stopwatch.createUnstarted();
     private static final GameRules TEST_GAME_RULES = Util.make(new GameRules(), p_351714_ -> {
-        p_351714_.getRule(GameRules.RULE_DOMOBSPAWNING).set(false, null);
-        p_351714_.getRule(GameRules.RULE_WEATHER_CYCLE).set(false, null);
-        p_351714_.getRule(GameRules.RULE_RANDOMTICKING).set(0, null);
-        p_351714_.getRule(GameRules.RULE_DOFIRETICK).set(false, null);
+        p_351714_.getRule(GameRules.RULE_DOMOBSPAWNING).set(false, (ServerLevel) null);
+        p_351714_.getRule(GameRules.RULE_WEATHER_CYCLE).set(false, (ServerLevel) null);
+        p_351714_.getRule(GameRules.RULE_RANDOMTICKING).set(0, (ServerLevel) null);
+        p_351714_.getRule(GameRules.RULE_DOFIRETICK).set(false, (ServerLevel) null);
     });
     private static final WorldOptions WORLD_OPTIONS = new WorldOptions(0L, false, false);
     @Nullable
@@ -151,19 +_,21 @@
     public boolean initServer() {
         this.setPlayerList(new PlayerList(this, this.registries(), this.playerDataStorage, 1) {
         });
-        this.loadLevel();
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
+        this.loadLevel(this.storageSource.getLevelId()); // CraftBukkit // Stellar - Fix compiler error
         ServerLevel serverlevel = this.overworld();
         this.testBatches = Lists.newArrayList(GameTestBatchFactory.fromTestFunction(this.testFunctions, serverlevel));
         serverlevel.setDefaultSpawnPos(this.spawnPos, 0.0F);
         int i = 20000000;
         serverlevel.setWeatherParameters(20000000, 20000000, false, false);
         LOGGER.info("Started game test server");
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
         return true;
     }
 
     @Override
-    public void tickServer(BooleanSupplier p_177619_) {
-        super.tickServer(p_177619_);
+    public void tickServer(BooleanSupplier shouldKeepTicking) {
+        super.tickServer(shouldKeepTicking);
         ServerLevel serverlevel = this.overworld();
         if (!this.haveTestsStarted()) {
             this.startTests(serverlevel);
@@ -301,5 +_,10 @@
     @Override
     public boolean isSingleplayerOwner(GameProfile p_177617_) {
         return false;
+    }
+
+    @Override
+    public org.bukkit.command.CommandSender getBukkitSender(net.minecraft.commands.CommandSourceStack wrapper) {
+        return null;
     }
 }
