--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -44,7 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.level.gameevent.GameEvent;import org.bukkit.event.entity.EntityDamageEvent;
 
 public class Armadillo extends Animal {
     public static final float BABY_SCALE = 0.6F;
@@ -139,7 +_,9 @@
         this.level().getProfiler().pop();
         if (this.isAlive() && !this.isBaby() && --this.scuteTime <= 0) {
             this.playSound(SoundEvents.ARMADILLO_SCUTE_DROP, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
+            this.forceDrops = true; // CraftBukkit
             this.spawnAtLocation(Items.ARMADILLO_SCUTE);
+            this.forceDrops = false; // CraftBukkit
             this.gameEvent(GameEvent.ENTITY_PLACE);
             this.scuteTime = this.pickNextScuteDropTime();
         }
@@ -274,33 +_,39 @@
     }
 
     @Override
-    public boolean hurt(DamageSource p_321611_, float p_321629_) {
+    public boolean hurt(DamageSource source, float amount) {
         if (this.isScared()) {
-            p_321629_ = (p_321629_ - 1.0F) / 2.0F;
+            amount = (amount - 1.0F) / 2.0F;
         }
 
-        return super.hurt(p_321611_, p_321629_);
+        return super.hurt(source, amount);
     }
 
     @Override
-    protected void actuallyHurt(DamageSource p_316204_, float p_316801_) {
-        super.actuallyHurt(p_316204_, p_316801_);
+    public boolean actuallyHurt(DamageSource source, float amount, EntityDamageEvent event) { // CraftBukkit start - void -> boolean
+        final var result = super.actuallyHurt(source, amount, event);
+        if (!result) {
+            return false;
+        }
+        // CraftBukkit end
         if (!this.isNoAi() && !this.isDeadOrDying()) {
-            if (p_316204_.getEntity() instanceof LivingEntity) {
+            if (source.getEntity() instanceof LivingEntity) {
                 this.getBrain().setMemoryWithExpiry(MemoryModuleType.DANGER_DETECTED_RECENTLY, true, 80L);
                 if (this.canStayRolledUp()) {
                     this.rollUp();
                 }
-            } else if (p_316204_.is(DamageTypeTags.PANIC_ENVIRONMENTAL_CAUSES)) {
+            } else if (source.is(DamageTypeTags.PANIC_ENVIRONMENTAL_CAUSES)) {
                 this.rollOut();
             }
         }
+
+        return true; // CraftBukkit
     }
 
     @Override
     public InteractionResult mobInteract(Player p_316559_, InteractionHand p_316119_) {
         ItemStack itemstack = p_316559_.getItemInHand(p_316119_);
-        if (itemstack.is(Items.BRUSH) && this.brushOffScute()) {
+        if (itemstack.canPerformAction(net.neoforged.neoforge.common.ItemAbilities.BRUSH_BRUSH) && this.brushOffScute()) {
             itemstack.hurtAndBreak(16, p_316559_, getSlotForHand(p_316119_));
             return InteractionResult.sidedSuccess(this.level().isClientSide);
         } else {
@@ -321,7 +_,9 @@
         if (this.isBaby()) {
             return false;
         } else {
+            this.forceDrops = true; // CraftBukkit
             this.spawnAtLocation(new ItemStack(Items.ARMADILLO_SCUTE));
+            this.forceDrops = false; // CraftBukkit
             this.gameEvent(GameEvent.ENTITY_INTERACT);
             this.playSound(SoundEvents.ARMADILLO_BRUSH);
             return true;
