--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -8,14 +_,42 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.entity.EnderChestBlockEntity;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.util.CraftLocation;
+import org.bukkit.inventory.InventoryHolder;
 
 public class PlayerEnderChestContainer extends SimpleContainer {
     @Nullable
     private EnderChestBlockEntity activeChest;
 
-    public PlayerEnderChestContainer() {
-        super(27);
-    }
+    // CraftBukkit start
+    private Player owner;
+
+    public InventoryHolder getBukkitOwner() {
+        return owner.getBukkitEntity();
+    }
+
+    public void setOwner(Player owner) {
+        this.owner = owner;
+    }
+
+    @Override
+    public Location getLocation() {
+        return this.activeChest != null ? CraftLocation.toBukkit(this.activeChest.getBlockPos(), this.activeChest.getLevel().getWorld()) : null;
+    }
+
+    public PlayerEnderChestContainer(Player owner) {
+        super(org.purpurmc.purpur.PurpurConfig.enderChestSixRows ? 54 : 27); // Purpur
+        this.owner = owner;
+        // CraftBukkit end
+    }
+
+    // Purpur start
+    @Override
+    public int getContainerSize() {
+        return owner.sixRowEnderchestSlotCount < 0 ? super.getContainerSize() : owner.sixRowEnderchestSlotCount;
+    }
+    // Purpur end
 
     public void setActiveChest(EnderChestBlockEntity p_40106_) {
         this.activeChest = p_40106_;
