--- a/net/minecraft/core/Direction.java
+++ b/net/minecraft/core/Direction.java
@@ -27,7 +_,7 @@
 import org.joml.Vector3f;
 import org.joml.Vector4f;
 
-public enum Direction implements StringRepresentable {
+public enum Direction implements StringRepresentable, ca.spottedleaf.moonrise.patches.collisions.util.CollisionDirection { // Paper - optimise collisions
     DOWN(0, 1, -1, "down", Direction.AxisDirection.NEGATIVE, Direction.Axis.Y, new Vec3i(0, -1, 0)),
     UP(1, 0, -1, "up", Direction.AxisDirection.POSITIVE, Direction.Axis.Y, new Vec3i(0, 1, 0)),
     NORTH(2, 3, 2, "north", Direction.AxisDirection.NEGATIVE, Direction.Axis.Z, new Vec3i(0, 0, -1)),
@@ -55,6 +_,39 @@
         .sorted(Comparator.comparingInt(p_235683_ -> p_235683_.data2d))
         .toArray(Direction[]::new);
 
+    // Paper start - optimise collisions
+    private static final int RANDOM_OFFSET = 2017601568;
+    private Direction opposite;
+    private Quaternionf rotation;
+    private int id;
+    private int stepX;
+    private int stepY;
+    private int stepZ;
+
+    private Quaternionf getRotationUncached() {
+        return switch ((Direction)(Object)this) {
+            case DOWN -> new Quaternionf().rotationX(3.1415927F);
+            case UP -> new Quaternionf();
+            case NORTH -> new Quaternionf().rotationXYZ(1.5707964F, 0.0F, 3.1415927F);
+            case SOUTH -> new Quaternionf().rotationX(1.5707964F);
+            case WEST -> new Quaternionf().rotationXYZ(1.5707964F, 0.0F, 1.5707964F);
+            case EAST -> new Quaternionf().rotationXYZ(1.5707964F, 0.0F, -1.5707964F);
+            default -> throw new IllegalStateException();
+        };
+    }
+
+    @Override
+    public final int moonrise$uniqueId() {
+        return this.id;
+    }
+    // Paper end - optimise collisions
+
+    // Stellar start - lithium: math.fast_blockpos - Avoid indirection to aid inlining
+    private final int offsetX;
+    private final int offsetY;
+    private final int offsetZ;
+    // Stellar end - lithium: math.fast_blockpos - Avoid indirection to aid inlining
+
     private Direction(
         int p_122356_, int p_122357_, int p_122358_, String p_122359_, Direction.AxisDirection p_122360_, Direction.Axis p_122361_, Vec3i p_122362_
     ) {
@@ -65,6 +_,12 @@
         this.axis = p_122361_;
         this.axisDirection = p_122360_;
         this.normal = p_122362_;
+
+        // Stellar start - lithium: math.fast_blockpos - Avoid indirection to aid inlining
+        this.offsetX = p_122362_.getX();
+        this.offsetY = p_122362_.getY();
+        this.offsetZ = p_122362_.getZ();
+        // Stellar end - lithium: math.fast_blockpos - Avoid indirection to aid inlining
     }
 
     public static Direction[] orderedByNearest(Entity p_122383_) {
@@ -117,14 +_,13 @@
     }
 
     public Quaternionf getRotation() {
-        return switch (this) {
-            case DOWN -> new Quaternionf().rotationX((float) Math.PI);
-            case UP -> new Quaternionf();
-            case NORTH -> new Quaternionf().rotationXYZ((float) (Math.PI / 2), 0.0F, (float) Math.PI);
-            case SOUTH -> new Quaternionf().rotationX((float) (Math.PI / 2));
-            case WEST -> new Quaternionf().rotationXYZ((float) (Math.PI / 2), 0.0F, (float) (Math.PI / 2));
-            case EAST -> new Quaternionf().rotationXYZ((float) (Math.PI / 2), 0.0F, (float) (-Math.PI / 2));
-        };
+        // Paper start - optimise collisions
+        try {
+            return (Quaternionf)this.rotation.clone();
+        } catch (final CloneNotSupportedException ex) {
+            throw new InternalError(ex);
+        }
+        // Paper end - optimise collisions
     }
 
     public int get3DDataValue() {
@@ -148,7 +_,9 @@
     }
 
     public Direction getOpposite() {
-        return from3DDataValue(this.oppositeIndex);
+        // Stellar start - lithium: math.fast_util - Avoid indirection to aid inlining
+        return this.opposite // Paper - optimise collisions
+        // Stellar end - lithium: math.fast_util - Avoid indirection to aid inlining
     }
 
     public Direction getClockWise(Direction.Axis p_175363_) {
@@ -227,16 +_,16 @@
         };
     }
 
-    public int getStepX() {
-        return this.normal.getX();
-    }
-
-    public int getStepY() {
-        return this.normal.getY();
-    }
-
-    public int getStepZ() {
-        return this.normal.getZ();
+    public final int getStepX() { // Stellar - Perf: Final for inline
+        return this.offsetX; // Stellar - lithium: math.fast_blockpos - Avoid indirection to aid inlining
+    }
+
+    public final int getStepY() { // Stellar - Perf: Final for inline
+        return this.offsetY; // Stellar - lithium: math.fast_blockpos - Avoid indirection to aid inlining
+    }
+
+    public final int getStepZ() { // Stellar - Perf: Final for inline
+        return this.offsetZ; // Stellar - lithium: math.fast_blockpos - Avoid indirection to aid inlining
     }
 
     public Vector3f step() {
@@ -310,7 +_,9 @@
     }
 
     public static Direction getRandom(RandomSource p_235673_) {
-        return Util.getRandom(VALUES, p_235673_);
+        // Stellar - lithium: math.fast_util - Avoid expensive array/modulo operations
+        return VALUES[p_235673_.nextInt(VALUES.length)];
+        // Stellar end - lithium: math.fast_util - Avoid expensive array/modulo operations
     }
 
     public static Direction getNearest(double p_122367_, double p_122368_, double p_122369_) {
@@ -322,7 +_,7 @@
         float f = Float.MIN_VALUE;
 
         for (Direction direction1 : VALUES) {
-            float f1 = p_122373_ * (float)direction1.normal.getX() + p_122374_ * (float)direction1.normal.getY() + p_122375_ * (float)direction1.normal.getZ();
+            float f1 = p_122373_ * (float)direction1.offsetX + p_122374_ * (float)direction1.offsetY + p_122375_ * (float)direction1.offsetZ; // Stellar - lithium: math.fast_blockpos - Avoid indirection to aid inlining
             if (f1 > f) {
                 f = f1;
                 direction = direction1;
@@ -368,7 +_,7 @@
         float f = p_122371_ * (float) (Math.PI / 180.0);
         float f1 = -Mth.sin(f);
         float f2 = Mth.cos(f);
-        return (float)this.normal.getX() * f1 + (float)this.normal.getZ() * f2 > 0.0F;
+        return (float)this.offsetX * f1 + (float)this.offsetZ * f2 > 0.0F; // Stellar - lithium: math.fast_blockpos - Avoid indirection to aid inlining
     }
 
     public static enum Axis implements StringRepresentable, Predicate<Direction> {
@@ -532,4 +_,17 @@
             return this.faces.length;
         }
     }
+
+    // Paper start - optimise collisions
+    static {
+        for (final Direction direction : VALUES) {
+            ((Direction)(Object)direction).opposite = from3DDataValue(((Direction)(Object)direction).oppositeIndex);
+            ((Direction)(Object)direction).rotation = ((Direction)(Object)direction).getRotationUncached();
+            ((Direction)(Object)direction).id = it.unimi.dsi.fastutil.HashCommon.murmurHash3(it.unimi.dsi.fastutil.HashCommon.murmurHash3(direction.ordinal() + RANDOM_OFFSET) + RANDOM_OFFSET);
+            ((Direction)(Object)direction).stepX = ((Direction)(Object)direction).normal.getX();
+            ((Direction)(Object)direction).stepY = ((Direction)(Object)direction).normal.getY();
+            ((Direction)(Object)direction).stepZ = ((Direction)(Object)direction).normal.getZ();
+        }
+    }
+    // Paper end - optimise collisions
 }
