--- a/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ b/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -236,4 +_,24 @@
             throw new RuntimeException(String.format(Locale.ROOT, "Unknown dimension type : %s", p_71332_));
         }
     }
+
+    public static LegacyStructureDataHandler getLegacyStructureHandlerCB(ResourceKey<net.minecraft.world.level.dimension.LevelStem> world, @Nullable DimensionDataStorage persistentStateManager) { // CraftBukkit
+        if (world == net.minecraft.world.level.dimension.LevelStem.OVERWORLD) { // CraftBukkit
+            return new LegacyStructureDataHandler(
+                    persistentStateManager,
+                    ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"),
+                    ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument")
+            );
+        } else {
+            if (world == net.minecraft.world.level.dimension.LevelStem.NETHER) { // CraftBukkit
+                List<String> list1 = ImmutableList.of("Fortress");
+                return new LegacyStructureDataHandler(persistentStateManager, list1, list1);
+            } else if (world == net.minecraft.world.level.dimension.LevelStem.END) { // CraftBukkit
+                List<String> list = ImmutableList.of("EndCity");
+                return new LegacyStructureDataHandler(persistentStateManager, list, list);
+            } else {
+                throw new RuntimeException(String.format(Locale.ROOT, "Unknown dimension type : %s", world));
+            }
+        }
+    }
 }
