--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -81,7 +_,7 @@
     public static final Codec<ChunkGenerator> CODEC = BuiltInRegistries.CHUNK_GENERATOR
         .byNameCodec()
         .dispatchStable(ChunkGenerator::codec, Function.identity());
-    protected final BiomeSource biomeSource;
+    protected BiomeSource biomeSource;
     private final Supplier<List<FeatureSorter.StepFeatureData>> featuresPerStep;
     public final Function<Holder<Biome>, BiomeGenerationSettings> generationSettingsGetter;
 
@@ -92,19 +_,24 @@
     public ChunkGenerator(BiomeSource p_255838_, Function<Holder<Biome>, BiomeGenerationSettings> p_256216_) {
         this.biomeSource = p_255838_;
         this.generationSettingsGetter = p_256216_;
-        this.featuresPerStep = Suppliers.memoize(
+        this.featuresPerStep = net.neoforged.neoforge.common.util.Lazy.of(
             () -> FeatureSorter.buildFeaturesPerStep(List.copyOf(p_255838_.possibleBiomes()), p_223216_ -> p_256216_.apply(p_223216_).features(), true)
         );
     }
 
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public void refreshFeaturesPerStep() {
+        ((net.neoforged.neoforge.common.util.Lazy<?>) this.featuresPerStep).invalidate();
+    }
+
     public void validate() {
         this.featuresPerStep.get();
     }
 
     protected abstract MapCodec<? extends ChunkGenerator> codec();
 
-    public ChunkGeneratorStructureState createState(HolderLookup<StructureSet> p_256405_, RandomState p_256101_, long p_256018_) {
-        return ChunkGeneratorStructureState.createForNormal(p_256101_, p_256018_, this.biomeSource, p_256405_);
+    public ChunkGeneratorStructureState createState(HolderLookup<StructureSet> p_256405_, RandomState p_256101_, long p_256018_, org.spigotmc.SpigotWorldConfig conf) { // Spigot
+        return ChunkGeneratorStructureState.createForNormal(p_256101_, p_256018_, this.biomeSource, p_256405_, conf); // Spigot
     }
 
     public Optional<ResourceKey<MapCodec<? extends ChunkGenerator>>> getTypeNameForDataFixer() {
@@ -115,7 +_,7 @@
         return CompletableFuture.supplyAsync(Util.wrapThreadWithTaskName("init_biomes", () -> {
             p_223163_.fillBiomesFromNoise(this.biomeSource, p_223160_.sampler());
             return p_223163_;
-        }), Util.backgroundExecutor());
+        }), Runnable::run); // Paper - rewrite chunk system
     }
 
     public abstract void applyCarvers(
@@ -132,6 +_,24 @@
     public Pair<BlockPos, Holder<Structure>> findNearestMapStructure(
         ServerLevel p_223038_, HolderSet<Structure> p_223039_, BlockPos p_223040_, int p_223041_, boolean p_223042_
     ) {
+        // Paper start - StructuresLocateEvent
+        final org.bukkit.World bukkitWorld = p_223038_.getWorld();
+        final org.bukkit.Location origin = io.papermc.paper.util.MCUtil.toLocation(p_223038_, p_223040_);
+        final List<org.bukkit.generator.structure.Structure> apiStructures = p_223039_.stream().map(Holder::value).map(nms -> org.bukkit.craftbukkit.generator.structure.CraftStructure.minecraftToBukkit(nms)).toList();
+        if (!apiStructures.isEmpty()) {
+            final io.papermc.paper.event.world.StructuresLocateEvent event = new io.papermc.paper.event.world.StructuresLocateEvent(bukkitWorld, origin, apiStructures, p_223041_, p_223042_);
+            if (!event.callEvent()) {
+                return null;
+            }
+            if (event.getResult() != null) {
+                return Pair.of(io.papermc.paper.util.MCUtil.toBlockPos(event.getResult().pos()), p_223038_.registryAccess().registryOrThrow(Registries.STRUCTURE).wrapAsHolder(org.bukkit.craftbukkit.generator.structure.CraftStructure.bukkitToMinecraft(event.getResult().structure())));
+            }
+            p_223040_ = io.papermc.paper.util.MCUtil.toBlockPosition(event.getOrigin());
+            p_223041_ = event.getRadius();
+            p_223042_ = event.shouldFindUnexplored();
+            p_223039_ = HolderSet.direct(api -> p_223038_.registryAccess().registryOrThrow(Registries.STRUCTURE).wrapAsHolder(org.bukkit.craftbukkit.generator.structure.CraftStructure.bukkitToMinecraft(api)), event.getStructures());
+        }
+        // Paper end
         ChunkGeneratorStructureState chunkgeneratorstructurestate = p_223038_.getChunkSource().getGeneratorState();
         Map<StructurePlacement, Set<Holder<Structure>>> map = new Object2ObjectArrayMap<>();
 
@@ -227,6 +_,8 @@
             BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
             for (ChunkPos chunkpos : list) {
+                if (!p_223183_.paperConfig().environment.locateStructuresOutsideWorldBorder && !p_223183_.getWorldBorder().isChunkInBounds(chunkpos.x, chunkpos.z)) { continue; } // Paper - Bound treasure maps to world border
+
                 blockpos$mutableblockpos.set(SectionPos.sectionToBlockCoord(chunkpos.x, 8), 32, SectionPos.sectionToBlockCoord(chunkpos.z, 8));
                 double d1 = blockpos$mutableblockpos.distSqr(p_223185_);
                 boolean flag = pair == null || d1 < d0;
@@ -293,7 +_,7 @@
                     return Pair.of(p_223203_.getLocatePos(p_223204_), holder);
                 }
 
-                ChunkAccess chunkaccess = p_223200_.getChunk(p_223204_.x, p_223204_.z, ChunkStatus.STRUCTURE_STARTS);
+                ChunkAccess chunkaccess = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevelReader)p_223200_).moonrise$syncLoadNonFull(p_223204_.x, p_223204_.z, ChunkStatus.STRUCTURE_STARTS); // Paper - rewrite chunk system
                 StructureStart structurestart = p_223201_.getStartForStructure(SectionPos.bottomOf(chunkaccess), holder.value(), chunkaccess);
                 if (structurestart != null && structurestart.isValid() && (!p_223202_ || tryAddReference(p_223201_, structurestart))) {
                     return Pair.of(p_223203_.getLocatePos(structurestart.getChunkPos()), holder);
@@ -313,99 +_,171 @@
         }
     }
 
-    public void applyBiomeDecoration(WorldGenLevel p_223087_, ChunkAccess p_223088_, StructureManager p_223089_) {
-        ChunkPos chunkpos = p_223088_.getPos();
-        if (!SharedConstants.debugVoidTerrain(chunkpos)) {
-            SectionPos sectionpos = SectionPos.of(chunkpos, p_223087_.getMinSection());
-            BlockPos blockpos = sectionpos.origin();
-            Registry<Structure> registry = p_223087_.registryAccess().registryOrThrow(Registries.STRUCTURE);
-            Map<Integer, List<Structure>> map = registry.stream().collect(Collectors.groupingBy(p_223103_ -> p_223103_.step().ordinal()));
-            List<FeatureSorter.StepFeatureData> list = this.featuresPerStep.get();
-            WorldgenRandom worldgenrandom = new WorldgenRandom(new XoroshiroRandomSource(RandomSupport.generateUniqueSeed()));
-            long i = worldgenrandom.setDecorationSeed(p_223087_.getSeed(), blockpos.getX(), blockpos.getZ());
-            Set<Holder<Biome>> set = new ObjectArraySet<>();
-            ChunkPos.rangeClosed(sectionpos.chunk(), 1).forEach(p_223093_ -> {
-                ChunkAccess chunkaccess = p_223087_.getChunk(p_223093_.x, p_223093_.z);
-
-                for (LevelChunkSection levelchunksection : chunkaccess.getSections()) {
-                    levelchunksection.getBiomes().getAll(set::add);
+    public void addVanillaDecorations(WorldGenLevel generatoraccessseed, ChunkAccess ichunkaccess, StructureManager structuremanager) { // CraftBukkit
+        ChunkPos chunkcoordintpair = ichunkaccess.getPos();
+
+        if (!SharedConstants.debugVoidTerrain(chunkcoordintpair)) {
+            SectionPos sectionposition = SectionPos.of(chunkcoordintpair, generatoraccessseed.getMinSection());
+            BlockPos blockposition = sectionposition.origin();
+            Registry<Structure> iregistry = generatoraccessseed.registryAccess().registryOrThrow(Registries.STRUCTURE);
+            Map<Integer, List<Structure>> map = (Map) iregistry.stream().collect(Collectors.groupingBy((structure) -> {
+                return structure.step().ordinal();
+            }));
+            List<FeatureSorter.StepFeatureData> list = (List) this.featuresPerStep.get();
+            WorldgenRandom seededrandom = new WorldgenRandom(new XoroshiroRandomSource(RandomSupport.generateUniqueSeed()));
+            long i = seededrandom.setDecorationSeed(generatoraccessseed.getSeed(), blockposition.getX(), blockposition.getZ());
+            Set<Holder<Biome>> set = new ObjectArraySet();
+
+            ChunkPos.rangeClosed(sectionposition.chunk(), 1).forEach((chunkcoordintpair1) -> {
+                ChunkAccess ichunkaccess1 = generatoraccessseed.getChunk(chunkcoordintpair1.x, chunkcoordintpair1.z);
+                LevelChunkSection[] achunksection = ichunkaccess1.getSections();
+                int j = achunksection.length;
+
+                for (int k = 0; k < j; ++k) {
+                    LevelChunkSection chunksection = achunksection[k];
+                    PalettedContainerRO<Holder<Biome>> palettedcontainerro = chunksection.getBiomes(); // CraftBukkit - decompile error
+
+                    java.util.Objects.requireNonNull(set);
+                    palettedcontainerro.getAll(set::add);
                 }
+
             });
             set.retainAll(this.biomeSource.possibleBiomes());
             int j = list.size();
 
             try {
-                Registry<PlacedFeature> registry1 = p_223087_.registryAccess().registryOrThrow(Registries.PLACED_FEATURE);
-                int i1 = Math.max(GenerationStep.Decoration.values().length, j);
-
-                for (int k = 0; k < i1; k++) {
-                    int l = 0;
-                    if (p_223089_.shouldGenerateStructures()) {
-                        for (Structure structure : map.getOrDefault(k, Collections.emptyList())) {
-                            worldgenrandom.setFeatureSeed(i, l, k);
-                            Supplier<String> supplier = () -> registry.getResourceKey(structure).map(Object::toString).orElseGet(structure::toString);
+                Registry<PlacedFeature> iregistry1 = generatoraccessseed.registryAccess().registryOrThrow(Registries.PLACED_FEATURE);
+                int k = Math.max(GenerationStep.Decoration.values().length, j);
+
+                for (int l = 0; l < k; ++l) {
+                    int i1 = 0;
+                    java.util.Iterator iterator;
+                    CrashReportCategory crashreportsystemdetails;
+
+                    if (structuremanager.shouldGenerateStructures()) {
+                        List<Structure> list1 = (List) map.getOrDefault(l, Collections.emptyList());
+
+                        for (iterator = list1.iterator(); iterator.hasNext(); ++i1) {
+                            Structure structure = (Structure) iterator.next();
+
+                            seededrandom.setFeatureSeed(i, i1, l);
+                            Supplier<String> supplier = () -> {
+                                Optional optional = iregistry.getResourceKey(structure).map(Object::toString);
+
+                                java.util.Objects.requireNonNull(structure);
+                                return (String) optional.orElseGet(structure::toString);
+                            };
 
                             try {
-                                p_223087_.setCurrentlyGenerating(supplier);
-                                p_223089_.startsForStructure(sectionpos, structure)
-                                    .forEach(
-                                        p_223086_ -> p_223086_.placeInChunk(p_223087_, p_223089_, this, worldgenrandom, getWritableArea(p_223088_), chunkpos)
-                                    );
+                                generatoraccessseed.setCurrentlyGenerating(supplier);
+                                structuremanager.startsForStructure(sectionposition, structure).forEach((structurestart) -> {
+                                    structurestart.placeInChunk(generatoraccessseed, structuremanager, this, seededrandom, ChunkGenerator.getWritableArea(ichunkaccess), chunkcoordintpair);
+                                });
                             } catch (Exception exception) {
-                                CrashReport crashreport1 = CrashReport.forThrowable(exception, "Feature placement");
-                                crashreport1.addCategory("Feature").setDetail("Description", supplier::get);
-                                throw new ReportedException(crashreport1);
-                            }
+                                CrashReport crashreport = CrashReport.forThrowable(exception, "Feature placement");
 
-                            l++;
+                                crashreportsystemdetails = crashreport.addCategory("Feature");
+                                java.util.Objects.requireNonNull(supplier);
+                                crashreportsystemdetails.setDetail("Description", supplier::get);
+                                throw new ReportedException(crashreport);
+                            }
                         }
                     }
 
-                    if (k < j) {
-                        IntSet intset = new IntArraySet();
-
-                        for (Holder<Biome> holder : set) {
-                            List<HolderSet<PlacedFeature>> list1 = this.generationSettingsGetter.apply(holder).features();
-                            if (k < list1.size()) {
-                                HolderSet<PlacedFeature> holderset = list1.get(k);
-                                FeatureSorter.StepFeatureData featuresorter$stepfeaturedata1 = list.get(k);
-                                holderset.stream()
-                                    .map(Holder::value)
-                                    .forEach(p_223174_ -> intset.add(featuresorter$stepfeaturedata1.indexMapping().applyAsInt(p_223174_)));
+                    if (l < j) {
+                        IntArraySet intarrayset = new IntArraySet();
+
+                        iterator = set.iterator();
+
+                        while (iterator.hasNext()) {
+                            Holder<Biome> holder = (Holder) iterator.next();
+                            List<HolderSet<PlacedFeature>> list2 = ((BiomeGenerationSettings) this.generationSettingsGetter.apply(holder)).features();
+
+                            if (l < list2.size()) {
+                                HolderSet<PlacedFeature> holderset = (HolderSet) list2.get(l);
+                                FeatureSorter.StepFeatureData featuresorter_b = (FeatureSorter.StepFeatureData) list.get(l);
+
+                                holderset.stream().map(Holder::value).forEach((placedfeature) -> {
+                                    intarrayset.add(featuresorter_b.indexMapping().applyAsInt(placedfeature));
+                                });
                             }
                         }
 
-                        int j1 = intset.size();
-                        int[] aint = intset.toIntArray();
+                        int j1 = intarrayset.size();
+                        int[] aint = intarrayset.toIntArray();
+
                         Arrays.sort(aint);
-                        FeatureSorter.StepFeatureData featuresorter$stepfeaturedata = list.get(k);
+                        FeatureSorter.StepFeatureData featuresorter_b1 = (FeatureSorter.StepFeatureData) list.get(l);
 
-                        for (int k1 = 0; k1 < j1; k1++) {
+                        for (int k1 = 0; k1 < j1; ++k1) {
                             int l1 = aint[k1];
-                            PlacedFeature placedfeature = featuresorter$stepfeaturedata.features().get(l1);
-                            Supplier<String> supplier1 = () -> registry1.getResourceKey(placedfeature).map(Object::toString).orElseGet(placedfeature::toString);
-                            worldgenrandom.setFeatureSeed(i, l1, k);
+                            PlacedFeature placedfeature = (PlacedFeature) featuresorter_b1.features().get(l1);
+                            Supplier<String> supplier1 = () -> {
+                                Optional optional = iregistry1.getResourceKey(placedfeature).map(Object::toString);
+
+                                java.util.Objects.requireNonNull(placedfeature);
+                                return (String) optional.orElseGet(placedfeature::toString);
+                            };
+
+                            // Paper start - Configurable feature seeds; change populationSeed used in random
+                            long featurePopulationSeed = i;
+                            final long configFeatureSeed = generatoraccessseed.getMinecraftWorld().paperConfig().featureSeeds.features.getLong(placedfeature.feature());
+                            if (configFeatureSeed != -1) {
+                                featurePopulationSeed = seededrandom.setDecorationSeed(configFeatureSeed, blockposition.getX(), blockposition.getZ()); // See seededrandom.setDecorationSeed from above
+                            }
+                            seededrandom.setFeatureSeed(featurePopulationSeed, l1, l);
+                            // Paper end - Configurable feature seeds
 
                             try {
-                                p_223087_.setCurrentlyGenerating(supplier1);
-                                placedfeature.placeWithBiomeCheck(p_223087_, this, worldgenrandom, blockpos);
+                                generatoraccessseed.setCurrentlyGenerating(supplier1);
+                                placedfeature.placeWithBiomeCheck(generatoraccessseed, this, seededrandom, blockposition);
                             } catch (Exception exception1) {
-                                CrashReport crashreport2 = CrashReport.forThrowable(exception1, "Feature placement");
-                                crashreport2.addCategory("Feature").setDetail("Description", supplier1::get);
-                                throw new ReportedException(crashreport2);
+                                CrashReport crashreport1 = CrashReport.forThrowable(exception1, "Feature placement");
+
+                                crashreportsystemdetails = crashreport1.addCategory("Feature");
+                                java.util.Objects.requireNonNull(supplier1);
+                                crashreportsystemdetails.setDetail("Description", supplier1::get);
+                                throw new ReportedException(crashreport1);
                             }
                         }
                     }
                 }
 
-                p_223087_.setCurrentlyGenerating(null);
+                generatoraccessseed.setCurrentlyGenerating((Supplier) null);
             } catch (Exception exception2) {
-                CrashReport crashreport = CrashReport.forThrowable(exception2, "Biome decoration");
-                crashreport.addCategory("Generation").setDetail("CenterX", chunkpos.x).setDetail("CenterZ", chunkpos.z).setDetail("Decoration Seed", i);
-                throw new ReportedException(crashreport);
-            }
-        }
-    }
+                CrashReport crashreport2 = CrashReport.forThrowable(exception2, "Biome decoration");
+
+                crashreport2.addCategory("Generation").setDetail("CenterX", (Object) chunkcoordintpair.x).setDetail("CenterZ", (Object) chunkcoordintpair.z).setDetail("Decoration Seed", (Object) i);
+                throw new ReportedException(crashreport2);
+            }
+        }
+    }
+
+    // CraftBukkit start
+    public void applyBiomeDecoration(WorldGenLevel world, ChunkAccess chunk, StructureManager structureAccessor) {
+        this.applyBiomeDecoration(world, chunk, structureAccessor, true);
+    }
+
+    public void applyBiomeDecoration(WorldGenLevel pLevel, ChunkAccess pChunk, StructureManager pStructureManager, boolean vanilla) {
+        if (vanilla) {
+            this.addVanillaDecorations(pLevel, pChunk, pStructureManager);
+        }
+        org.bukkit.World world = pLevel.getMinecraftWorld().getWorld();
+        // only call when a populator is present (prevents unnecessary entity conversion)
+        if (!world.getPopulators().isEmpty()) {
+            org.bukkit.craftbukkit.generator.CraftLimitedRegion limitedRegion = new org.bukkit.craftbukkit.generator.CraftLimitedRegion(pLevel, pChunk.getPos());
+            int x = pChunk.getPos().x;
+            int z = pChunk.getPos().z;
+            for (org.bukkit.generator.BlockPopulator populator : world.getPopulators()) {
+                WorldgenRandom seededrandom = new WorldgenRandom(new LegacyRandomSource(pLevel.getSeed()));
+                seededrandom.setDecorationSeed(pLevel.getSeed(), x, z);
+                populator.populate(world, new org.bukkit.craftbukkit.util.RandomSourceWrapper.RandomWrapper(seededrandom), x, z, limitedRegion);
+            }
+            limitedRegion.saveEntities();
+            limitedRegion.breakLink();
+        }
+    }
+    // CraftBukkit end
 
     private static BoundingBox getWritableArea(ChunkAccess p_187718_) {
         ChunkPos chunkpos = p_187718_.getPos();
@@ -429,6 +_,10 @@
         return this.biomeSource;
     }
 
+    public void setBiomeSource(BiomeSource biomeSource) {
+        this.biomeSource = biomeSource;
+    }
+
     public abstract int getGenDepth();
 
     public WeightedRandomList<MobSpawnSettings.SpawnerData> getMobsAt(
@@ -551,6 +_,22 @@
             p_223107_, this, this.biomeSource, p_223108_, p_223109_, p_223110_, p_223112_, i, p_223111_, predicate
         );
         if (structurestart.isValid()) {
+            if (org.spigotmc.AsyncCatcher.catchAsync()) {
+                // CraftBukkit start
+                BoundingBox box = structurestart.getBoundingBox();
+                org.bukkit.event.world.AsyncStructureSpawnEvent event = new org.bukkit.event.world.AsyncStructureSpawnEvent(
+                        p_223106_.level.getMinecraftWorld().getWorld(),
+                        org.bukkit.craftbukkit.generator.structure.CraftStructure.minecraftToBukkit(structure),
+                        new org.bukkit.util.BoundingBox(box.minX(), box.minY(), box.minZ(), box.maxX(), box.maxY(), box.maxZ()),
+                        p_223112_.x,
+                        p_223112_.z
+                );
+                org.bukkit.Bukkit.getPluginManager().callEvent(event);
+                if (event.isCancelled()) {
+                    return true;
+                }
+                // CraftBukkit end
+            }
             p_223106_.setStartForStructure(p_223113_, structure, structurestart, p_223111_);
             return true;
         } else {
