--- a/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
@@ -20,6 +_,13 @@
         this.status = p_272864_;
         this.connection = p_273586_;
     }
+    // Neo start
+    public ServerStatusPacketListenerImpl(ServerStatus p_272864_, Connection p_273586_, @org.jetbrains.annotations.Nullable String statusCache) {
+        this.status = p_272864_;
+        this.connection = p_273586_;
+        // Stellar - unused field
+    }
+    // Neo end
 
     @Override
     public void onDisconnect(DisconnectionDetails p_350895_) {
@@ -36,7 +_,10 @@
             this.connection.disconnect(DISCONNECT_REASON);
         } else {
             this.hasRequestedStatus = true;
-            this.connection.send(new ClientboundStatusResponsePacket(this.status));
+            // Paper start - Replace everything
+            if (net.minecraft.server.MinecraftServer.getServer().getStatus().version().isEmpty()) return; // Purpur - do not respond to pings before we know the protocol version
+            com.destroystokyo.paper.network.StandardPaperServerListPingEventImpl.processRequest(net.minecraft.server.MinecraftServer.getServer(), this.connection);
+            // Paper end
         }
     }
 
