--- a/net/minecraft/util/random/WeightedRandomList.java
+++ b/net/minecraft/util/random/WeightedRandomList.java
@@ -6,15 +_,20 @@
 import java.util.Objects;
 import java.util.Optional;
 import javax.annotation.Nullable;
+
 import net.minecraft.util.RandomSource;
 
 public class WeightedRandomList<E extends WeightedEntry> {
     private final int totalWeight;
     private final ImmutableList<E> items;
+    private final List<E> entryHashList;
 
     WeightedRandomList(List<? extends E> p_146327_) {
         this.items = ImmutableList.copyOf(p_146327_);
         this.totalWeight = WeightedRandom.getTotalWeight(p_146327_);
+        // Stellar - lithium: Uses custom hashset/list combination for faster mob spawn checks
+        // We are using reference equality here, because all vanilla implementations of Weighted use reference equality
+        this.entryHashList = this.items.size() > 4 ? java.util.Collections.unmodifiableList(new net.caffeinemc.mods.lithium.common.util.collections.HashedReferenceList<>(this.items)) : this.items;
     }
 
     public static <E extends WeightedEntry> WeightedRandomList<E> create() {
@@ -44,7 +_,7 @@
     }
 
     public List<E> unwrap() {
-        return this.items;
+        return this.entryHashList; // Stellar - lithium: Uses custom hashset/list combination for faster mob spawn checks
     }
 
     public static <E extends WeightedEntry> Codec<WeightedRandomList<E>> codec(Codec<E> p_146334_) {
