--- a/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/net/minecraft/world/item/crafting/RecipeManager.java
@@ -38,7 +_,7 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private final HolderLookup.Provider registries;
     public Multimap<RecipeType<?>, RecipeHolder<?>> byType = ImmutableMultimap.of();
-    private Map<ResourceLocation, RecipeHolder<?>> byName = ImmutableMap.of();
+    public Map<ResourceLocation, RecipeHolder<?>> byName = ImmutableMap.of();
     private boolean hasErrors;
 
     public RecipeManager(HolderLookup.Provider p_324137_) {
@@ -50,25 +_,47 @@
         this.hasErrors = false;
         Builder<RecipeType<?>, RecipeHolder<?>> builder = ImmutableMultimap.builder();
         com.google.common.collect.ImmutableMap.Builder<ResourceLocation, RecipeHolder<?>> builder1 = ImmutableMap.builder();
-        RegistryOps<JsonElement> registryops = this.registries.createSerializationContext(JsonOps.INSTANCE);
+        RegistryOps<JsonElement> registryops = this.makeConditionalOps(); // Neo: add condition context
 
         for (Entry<ResourceLocation, JsonElement> entry : p_44037_.entrySet()) {
             ResourceLocation resourcelocation = entry.getKey();
+            if (resourcelocation.getPath().startsWith("_")) continue; //Forge: filter anything beginning with "_" as it's used for metadata.
 
             try {
-                Recipe<?> recipe = Recipe.CODEC.parse(registryops, entry.getValue()).getOrThrow(JsonParseException::new);
+                var decoded = Recipe.CONDITIONAL_CODEC.parse(registryops, entry.getValue()).getOrThrow(JsonParseException::new);
+                decoded.ifPresentOrElse(r -> {
+                Recipe<?> recipe = r.carrier();
                 RecipeHolder<?> recipeholder = new RecipeHolder<>(resourcelocation, recipe);
                 builder.put(recipe.getType(), recipeholder);
                 builder1.put(resourcelocation, recipeholder);
+                }, () -> {
+                    LOGGER.debug("Skipping loading recipe {} as its conditions were not met", resourcelocation);
+                });
             } catch (IllegalArgumentException | JsonParseException jsonparseexception) {
                 LOGGER.error("Parsing error loading recipe {}", resourcelocation, jsonparseexception);
             }
         }
 
-        this.byType = builder.build();
-        this.byName = builder1.build();
-        LOGGER.info("Loaded {} recipes", this.byType.size());
-    }
+        // CraftBukkit start - mutable
+        this.byType = com.google.common.collect.LinkedHashMultimap.create(builder.build());
+        this.byName = com.google.common.collect.Maps.newHashMap(builder1.build());
+        // CraftBukkit end
+        LOGGER.info("Loaded {} recipes", this.byName.size()); // Paper - Improve logging and errors; log correct number of recipes
+    }
+
+    // CraftBukkit start
+    public void addRecipe(RecipeHolder<?> recipe) {
+        org.spigotmc.AsyncCatcher.catchOp("Recipe Add"); // Spigot
+        Collection<RecipeHolder<?>> map = this.byType.get(recipe.value().getType()); // CraftBukkit
+
+        if (this.byName.containsKey(recipe.id())) {
+            throw new IllegalStateException("Duplicate recipe ignored with ID " + recipe.id());
+        } else {
+            map.add(recipe);
+            this.byName.put(recipe.id(), recipe);
+        }
+    }
+    // CraftBukkit end
 
     public boolean hadErrorsLoading() {
         return this.hasErrors;
@@ -88,13 +_,23 @@
     public <I extends RecipeInput, T extends Recipe<I>> Optional<RecipeHolder<T>> getRecipeFor(
         RecipeType<T> p_220249_, I p_345826_, Level p_220251_, @Nullable RecipeHolder<T> p_346407_
     ) {
+        // Paper start - fix upstream's complete screw up of checking last used recipe
         if (p_345826_.isEmpty()) {
             return Optional.empty();
+        } else if (p_346407_ != null && p_346407_.value().matches(p_345826_, p_220251_)) {
+            return Optional.of(p_346407_);
         } else {
-            return p_346407_ != null && p_346407_.value().matches(p_345826_, p_220251_)
-                ? Optional.of(p_346407_)
-                : this.byType(p_220249_).stream().filter(p_344413_ -> p_344413_.value().matches(p_345826_, p_220251_)).findFirst();
+            // Stellar start - Carpet-Fixes: Use optimized RecipeManager - Remove streams to be faster
+            RecipeHolder<T> lastRecipe = null;
+            for (RecipeHolder<T> recipeHolder : this.byType(p_220249_)) {
+                if (recipeHolder.value().matches(p_345826_, p_220251_)) {
+                    lastRecipe = recipeHolder;
+                }
+            }
+            return lastRecipe != null ? Optional.of(lastRecipe) : Optional.empty(); // CraftBukkit - SPIGOT-4638: last recipe gets priority
+            // Stellar end - Carpet-Fixes: Use optimized RecipeManager - Remove streams to be faster
         }
+        // Paper end
     }
 
     public <I extends RecipeInput, T extends Recipe<I>> List<RecipeHolder<T>> getAllRecipesFor(RecipeType<T> p_44014_) {
@@ -114,7 +_,12 @@
     }
 
     public <I extends RecipeInput, T extends Recipe<I>> NonNullList<ItemStack> getRemainingItemsFor(RecipeType<T> p_44070_, I p_345118_, Level p_44072_) {
-        Optional<RecipeHolder<T>> optional = this.getRecipeFor(p_44070_, p_345118_, p_44072_);
+        // Paper start - Perf: improve performance of mass crafts
+        return this.getRemainingItemsFor(p_44070_, p_345118_, p_44072_, null);
+    }
+    public <I extends RecipeInput, T extends Recipe<I>> NonNullList<ItemStack> getRemainingItemsFor(RecipeType<T> p_44070_, I p_345118_, Level p_44072_, @Nullable RecipeHolder<T> previousRecipe) {
+        Optional<RecipeHolder<T>> optional = this.getRecipeFor(p_44070_, p_345118_, p_44072_, previousRecipe);
+        // Paper end - Perf: improve performance of mass crafts
         if (optional.isPresent()) {
             return optional.get().value().getRemainingItems(p_345118_);
         } else {
@@ -194,4 +_,24 @@
     public interface CachedCheck<I extends RecipeInput, T extends Recipe<I>> {
         Optional<RecipeHolder<T>> getRecipeFor(I p_344938_, Level p_220281_);
     }
+
+    // CraftBukkit start
+    public boolean removeRecipe(ResourceLocation mcKey) {
+        org.spigotmc.AsyncCatcher.catchOp("Recipe Remove"); // Stellar - missed catch-op
+        java.util.Iterator<RecipeHolder<?>> iter = this.byType.values().iterator();
+        while (iter.hasNext()) {
+            RecipeHolder<?> recipe = iter.next();
+            if (recipe.id().equals(mcKey)) {
+                iter.remove();
+            }
+        }
+
+        return this.byName.remove(mcKey) != null;
+    }
+
+    public void clearRecipes() {
+        this.byType = com.google.common.collect.LinkedHashMultimap.create();
+        this.byName = com.google.common.collect.Maps.newHashMap();
+    }
+    // CraftBukkit end
 }
