--- a/net/minecraft/world/entity/ai/Brain.java
+++ b/net/minecraft/world/entity/ai/Brain.java
@@ -15,7 +_,8 @@
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.MapLike;
 import com.mojang.serialization.RecordBuilder;
-import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import it.unimi.dsi.fastutil.objects.*;
+
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
@@ -25,6 +_,7 @@
 import java.util.function.Supplier;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerLevel;
@@ -46,11 +_,11 @@
     static final Logger LOGGER = LogUtils.getLogger();
     private final Supplier<Codec<Brain<E>>> codec;
     private static final int SCHEDULE_UPDATE_DELAY = 20;
-    private final Map<MemoryModuleType<?>, Optional<? extends ExpirableValue<?>>> memories = Maps.newHashMap();
-    private final Map<SensorType<? extends Sensor<? super E>>, Sensor<? super E>> sensors = Maps.newLinkedHashMap();
+    private final Reference2ObjectMap<MemoryModuleType<?>, Optional<? extends ExpirableValue<?>>> memories = new Reference2ObjectOpenHashMap<>(); // Stellar - lithium: Use fastutil
+    private final Map<SensorType<? extends Sensor<? super E>>, Sensor<? super E>> sensors = new Reference2ReferenceLinkedOpenHashMap<>(); // Stellar - lithium: Use fastutil
     private final Map<Integer, Map<Activity, Set<BehaviorControl<? super E>>>> availableBehaviorsByPriority = Maps.newTreeMap();
     private Schedule schedule = Schedule.EMPTY;
-    private final Map<Activity, Set<Pair<MemoryModuleType<?>, MemoryStatus>>> activityRequirements = Maps.newHashMap();
+    private final Map<Activity, Set<Pair<MemoryModuleType<?>, MemoryStatus>>> activityRequirements = new Object2ObjectOpenHashMap<>(); // Stellar - lithium: Use fastutil
     private final Map<Activity, Set<MemoryModuleType<?>>> activityMemoriesToEraseWhenStopped = Maps.newHashMap();
     private Set<Activity> coreActivities = Sets.newHashSet();
     private final Set<Activity> activeActivities = Sets.newHashSet();
@@ -71,13 +_,22 @@
             (new MapCodec<Brain<E>>() {
                     @Override
                     public <T> Stream<T> keys(DynamicOps<T> p_22029_) {
-                        return p_21947_.stream()
-                            .flatMap(
-                                p_22020_ -> p_22020_.getCodec()
-                                        .map(p_258254_ -> BuiltInRegistries.MEMORY_MODULE_TYPE.getKey((MemoryModuleType<?>)p_22020_))
-                                        .stream()
-                            )
-                            .map(p_22018_ -> p_22029_.createString(p_22018_.toString()));
+                        // Stellar start - Leaf: Remove stream in Brain
+                        List<T> results = new java.util.ArrayList<>();
+
+                        for (final MemoryModuleType<?> memoryModuleType : p_21947_) {
+                            final Optional<?> codec = memoryModuleType.getCodec();
+
+                            if (codec.isPresent()) {
+                                final net.minecraft.resources.ResourceLocation resourceLocation = BuiltInRegistries.MEMORY_MODULE_TYPE.getKey(memoryModuleType);
+                                final T opsResult = p_22029_.createString(resourceLocation.toString());
+
+                                results.add(opsResult);
+                            }
+                        }
+
+                        return results.stream();
+                        // Stellar end - Leaf: Remove stream in Brain
                     }
 
                     @Override
@@ -113,7 +_,13 @@
                     }
 
                     public <T> RecordBuilder<T> encode(Brain<E> p_21985_, DynamicOps<T> p_21986_, RecordBuilder<T> p_21987_) {
-                        p_21985_.memories().forEach(p_22007_ -> p_22007_.serialize(p_21986_, p_21987_));
+                        // Stellar start - Leaf: Remove stream in Brain
+                        for (final Entry<MemoryModuleType<?>, Optional<? extends ExpirableValue<?>>> memory : p_21985_.memories.entrySet()) {
+                            final Brain.MemoryValue<?> result = Brain.MemoryValue.createUnchecked(memory.getKey(), memory.getValue());
+
+                            result.serialize(p_21986_, p_21987_);
+                        }
+                        // Stellar end - Leaf: Remove stream in Brain
                         return p_21987_;
                     }
                 })
@@ -155,7 +_,7 @@
     }
 
     Stream<Brain.MemoryValue<?>> memories() {
-        return this.memories.entrySet().stream().map(p_21929_ -> Brain.MemoryValue.createUnchecked(p_21929_.getKey(), p_21929_.getValue()));
+        return this.memories.entrySet().stream().map(p_21929_ -> Brain.MemoryValue.createUnchecked(p_21929_.getKey(), p_21929_.getValue())); // Stellar - Leaf: Remove stream in Brain - diff on change
     }
 
     public boolean hasMemoryValue(MemoryModuleType<?> p_21875_) {
@@ -195,7 +_,7 @@
     public <U> Optional<U> getMemory(MemoryModuleType<U> p_21953_) {
         Optional<? extends ExpirableValue<?>> optional = this.memories.get(p_21953_);
         if (optional == null) {
-            throw new IllegalStateException("Unregistered memory fetched: " + p_21953_);
+            return Optional.empty();
         } else {
             return (Optional<U>)optional.map(ExpirableValue::getValue);
         }
@@ -399,7 +_,9 @@
         return brain;
     }
 
+    private int behaviorTick = 0; // Pufferfish
     public void tick(ServerLevel p_21866_, E p_21867_) {
+        if (this.behaviorTick++ % p_21867_.activatedPriority == 0) return; // Pufferfish
         this.forgetOutdatedMemories();
         this.tickSensors(p_21866_, p_21867_);
         this.startEachNonRunningBehavior(p_21866_, p_21867_);
@@ -413,7 +_,7 @@
     }
 
     private void forgetOutdatedMemories() {
-        for (Entry<MemoryModuleType<?>, Optional<? extends ExpirableValue<?>>> entry : this.memories.entrySet()) {
+        for (Entry<MemoryModuleType<?>, Optional<? extends ExpirableValue<?>>> entry : this.memories.reference2ObjectEntrySet()) { // Stellar - lithium: Use fastutil
             if (entry.getValue().isPresent()) {
                 ExpirableValue<?> expirablevalue = (ExpirableValue<?>)entry.getValue().get();
                 if (expirablevalue.hasExpired()) {
