--- a/net/minecraft/world/level/gameevent/GameEventDispatcher.java
+++ b/net/minecraft/world/level/gameevent/GameEventDispatcher.java
@@ -21,6 +_,14 @@
     public void post(Holder<GameEvent> p_316151_, Vec3 p_250613_, GameEvent.Context p_251777_) {
         int i = p_316151_.value().notificationRadius();
         BlockPos blockpos = BlockPos.containing(p_250613_);
+        // CraftBukkit start
+        org.bukkit.event.world.GenericGameEvent event1 = new org.bukkit.event.world.GenericGameEvent(org.bukkit.craftbukkit.CraftGameEvent.minecraftToBukkit(p_316151_.value()), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(blockpos, this.level.getWorld()), (p_251777_.sourceEntity() == null) ? null : p_251777_.sourceEntity().getBukkitEntity(), i, !org.bukkit.Bukkit.isPrimaryThread());
+        this.level.getCraftServer().getPluginManager().callEvent(event1);
+        if (event1.isCancelled()) {
+            return;
+        }
+        i = event1.getRadius();
+        // CraftBukkit end
         int j = SectionPos.blockToSectionCoord(blockpos.getX() - i);
         int k = SectionPos.blockToSectionCoord(blockpos.getY() - i);
         int l = SectionPos.blockToSectionCoord(blockpos.getZ() - i);
@@ -39,7 +_,7 @@
 
         for (int l1 = j; l1 <= i1; l1++) {
             for (int i2 = l; i2 <= k1; i2++) {
-                ChunkAccess chunkaccess = this.level.getChunkSource().getChunkNow(l1, i2);
+                ChunkAccess chunkaccess = this.level.getChunkIfLoadedImmediately(l1, i2); // Paper - Use getChunkIfLoadedImmediately
                 if (chunkaccess != null) {
                     for (int j2 = k; j2 <= j1; j2++) {
                         flag |= chunkaccess.getListenerRegistry(j2)
