--- a/net/minecraft/advancements/critereon/TargetBlockTrigger.java
+++ b/net/minecraft/advancements/critereon/TargetBlockTrigger.java
@@ -18,6 +_,7 @@
     }
 
     public void trigger(ServerPlayer p_70212_, Entity p_70213_, Vec3 p_70214_, int p_70215_) {
+        if (dev.instellar.stellar.configuration.GlobalConfiguration.get().players.disableCriterionTrigger) return; // Stellar - disable criterion trigger config
         LootContext lootcontext = EntityPredicate.createContext(p_70212_, p_70213_);
         this.trigger(p_70212_, p_70224_ -> p_70224_.matches(lootcontext, p_70214_, p_70215_));
     }
