--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -24,7 +_,7 @@
 
     void setItem(int p_18944_, ItemStack p_18945_);
 
-    default int getMaxStackSize() {
+    default int getMaxStackSize() { // CraftBukkit // Stellar - modded container support
         return 99;
     }
 
@@ -91,4 +_,36 @@
             return level.getBlockEntity(blockpos) != p_272877_ ? false : p_272670_.canInteractWithBlock(blockpos, (double)p_320837_);
         }
     }
+
+    // CraftBukkit start // Stellar start - modded container support
+    default java.util.List<ItemStack> getContents() {
+        return java.util.Collections.emptyList();
+    }
+
+    default void onOpen(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+    }
+
+    default void onClose(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+    }
+
+    default java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return java.util.Collections.emptyList();
+    }
+
+    @org.jetbrains.annotations.Nullable org.bukkit.inventory.InventoryHolder getOwner(); // Paper - annotation
+
+    default void setMaxStackSize(int size) {
+    }
+
+    default org.bukkit.Location getLocation() {
+        if (this instanceof BlockEntity entity) {
+            BlockPos blockPos = entity.getBlockPos();
+            return new org.bukkit.Location(entity.getLevel().getWorld(), blockPos.getX(), blockPos.getY(), blockPos.getZ());
+        } else {
+            return null;
+        }
+    }
+
+    int MAX_STACK = 99;
+    // CraftBukkit end // Stellar end - modded container support
 }
