--- a/net/minecraft/network/protocol/game/ClientboundCommandsPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundCommandsPacket.java
@@ -242,8 +_,21 @@
         private static <A extends ArgumentType<?>, T extends ArgumentTypeInfo.Template<A>> void serializeCap(
             FriendlyByteBuf p_237663_, ArgumentTypeInfo<A, T> p_237664_, ArgumentTypeInfo.Template<A> p_237665_
         ) {
+            // Youer start
+            ResourceLocation key = BuiltInRegistries.COMMAND_ARGUMENT_TYPE.getKey(p_237664_);
+            if ((!io.papermc.paper.configuration.GlobalConfiguration.get().proxies.velocity.enabled && !org.spigotmc.SpigotConfig.bungee) || ((key != null) && (key.getNamespace().equals("minecraft") || key.getNamespace().equals("brigadier")))) {
+                p_237663_.writeVarInt(BuiltInRegistries.COMMAND_ARGUMENT_TYPE.getId(p_237664_));
+                p_237664_.serializeToNetwork((T) p_237665_, p_237663_);
+                return;
+            }
+            p_237663_.writeVarInt(-256);
             p_237663_.writeVarInt(BuiltInRegistries.COMMAND_ARGUMENT_TYPE.getId(p_237664_));
-            p_237664_.serializeToNetwork((T)p_237665_, p_237663_);
+            FriendlyByteBuf payload = new FriendlyByteBuf(io.netty.buffer.Unpooled.buffer());
+            p_237664_.serializeToNetwork((T) p_237665_, payload);
+            p_237663_.writeVarInt(payload.readableBytes());
+            p_237663_.writeBytes(payload);
+            payload.release();
+            // Youer end
         }
     }
 
