--- a/net/minecraft/world/level/levelgen/Beardifier.java
+++ b/net/minecraft/world/level/levelgen/Beardifier.java
@@ -37,13 +_,19 @@
         int j = p_223939_.getMinBlockZ();
         ObjectList<Beardifier.Rigid> objectlist = new ObjectArrayList<>(10);
         ObjectList<JigsawJunction> objectlist1 = new ObjectArrayList<>(32);
-        p_223938_.startsForStructure(p_223939_, p_223941_ -> p_223941_.terrainAdaptation() != TerrainAdjustment.NONE)
-            .forEach(
-                p_223936_ -> {
+        // Paper start - Perf: Remove streams from hot code
+        for (net.minecraft.world.level.levelgen.structure.StructureStart p_223936_ : p_223938_.startsForStructure(p_223939_, structure -> {
+            return structure.terrainAdaptation() != TerrainAdjustment.NONE;
+        })) { // Paper end - Perf: Remove streams from hot code
                     TerrainAdjustment terrainadjustment = p_223936_.getStructure().terrainAdaptation();
 
                     for (StructurePiece structurepiece : p_223936_.getPieces()) {
                         if (structurepiece.isCloseToChunk(p_223939_, 12)) {
+                            if (structurepiece instanceof net.neoforged.neoforge.common.world.PieceBeardifierModifier pieceBeardifierModifier) {
+                                if (pieceBeardifierModifier.getTerrainAdjustment() != TerrainAdjustment.NONE) {
+                                    objectlist.add(new Beardifier.Rigid(pieceBeardifierModifier.getBeardifierBox(), pieceBeardifierModifier.getTerrainAdjustment(), pieceBeardifierModifier.getGroundLevelDelta()));
+                                }
+                            } else
                             if (structurepiece instanceof PoolElementStructurePiece) {
                                 PoolElementStructurePiece poolelementstructurepiece = (PoolElementStructurePiece)structurepiece;
                                 StructureTemplatePool.Projection structuretemplatepool$projection = poolelementstructurepiece.getElement().getProjection();
@@ -68,7 +_,7 @@
                         }
                     }
                 }
-            );
+        // Paper end - Perf: Remove streams from hot code
         return new Beardifier(objectlist.iterator(), objectlist1.iterator());
     }
 
@@ -134,8 +_,14 @@
     }
 
     protected static double getBuryContribution(double p_340947_, double p_340921_, double p_341266_) {
-        double d0 = Mth.length(p_340947_, p_340921_, p_341266_);
-        return Mth.clampedMap(d0, 0.0, 6.0, 1.0, 0.0);
+        // Stellar start - C2ME: Optimize world gen math
+        double len = Math.sqrt(p_340947_ * p_340947_ + p_340921_ * p_340921_ + p_341266_ * p_341266_);
+        if (len > 6.0) {
+            return 0.0;
+        } else {
+            return 1.0 - len / 6.0;
+        }
+        // Stellar end - C2ME: Optimize world gen math
     }
 
     protected static double getBeardContribution(int p_223926_, int p_223927_, int p_223928_, int p_223929_) {
