--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -14,7 +_,7 @@
 import java.util.function.Predicate;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
-import net.minecraft.SharedConstants;
+import dev.instellar.stellar.configuration.GlobalConfiguration;import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
@@ -69,8 +_,10 @@
         Aquifer.FluidStatus aquifer$fluidstatus = new Aquifer.FluidStatus(-54, Blocks.LAVA.defaultBlockState());
         int i = p_249264_.seaLevel();
         Aquifer.FluidStatus aquifer$fluidstatus1 = new Aquifer.FluidStatus(i, p_249264_.defaultFluid());
-        Aquifer.FluidStatus aquifer$fluidstatus2 = new Aquifer.FluidStatus(DimensionType.MIN_Y * 2, Blocks.AIR.defaultBlockState());
-        return (p_224274_, p_224275_, p_224276_) -> p_224275_ < Math.min(-54, i) ? aquifer$fluidstatus : aquifer$fluidstatus1;
+        // Stellar start - C2ME: Optimize world gen math
+        final int min = Math.min(-54, i);
+        return (x, y, z) -> y < min ? aquifer$fluidstatus : aquifer$fluidstatus1;
+        // Stellar end - C2ME: Optimize world gen math
     }
 
     @Override
@@ -78,7 +_,7 @@
         return CompletableFuture.supplyAsync(Util.wrapThreadWithTaskName("init_biomes", () -> {
             this.doCreateBiomes(p_224300_, p_224299_, p_224301_, p_224302_);
             return p_224302_;
-        }), Util.backgroundExecutor());
+        }), Runnable::run); // Paper - rewrite chunk system
     }
 
     private void doCreateBiomes(Blender p_224292_, RandomState p_224293_, StructureManager p_224294_, ChunkAccess p_224295_) {
@@ -231,7 +_,7 @@
     @Override
     public void buildSurface(WorldGenRegion p_224232_, StructureManager p_224233_, RandomState p_224234_, ChunkAccess p_224235_) {
         if (!SharedConstants.debugVoidTerrain(p_224235_.getPos())) {
-            WorldGenerationContext worldgenerationcontext = new WorldGenerationContext(this, p_224232_);
+            WorldGenerationContext worldgenerationcontext = new WorldGenerationContext(this, p_224232_, p_224232_.getMinecraftWorld()); // Paper - Flat bedrock generator settings
             this.buildSurface(
                 p_224235_,
                 worldgenerationcontext,
@@ -288,7 +_,7 @@
         NoiseChunk noisechunk = p_224229_.getOrCreateNoiseChunk(p_224250_ -> this.createNoiseChunk(p_224250_, p_224228_, Blender.of(p_224224_), p_224226_));
         Aquifer aquifer = noisechunk.aquifer();
         CarvingContext carvingcontext = new CarvingContext(
-            this, p_224224_.registryAccess(), p_224229_.getHeightAccessorForGeneration(), noisechunk, p_224226_, this.settings.value().surfaceRule()
+            this, p_224224_.registryAccess(), p_224229_.getHeightAccessorForGeneration(), noisechunk, p_224226_, this.settings.value().surfaceRule(), p_224224_.getMinecraftWorld() // Paper - Flat bedrock generator settings
         );
         CarvingMask carvingmask = ((ProtoChunk)p_224229_).getOrCreateCarvingMask(p_224230_);
 
@@ -320,6 +_,8 @@
         }
     }
 
+    public static boolean USE_NOISIUM_WORLDGEN; // Stellar - Noisium: Optimize noise chunk generation
+
     @Override
     public CompletableFuture<ChunkAccess> fillFromNoise(Blender p_224313_, RandomState p_224314_, StructureManager p_224315_, ChunkAccess p_224316_) {
         NoiseSettings noisesettings = this.settings.value().noiseSettings().clampToHeightAccessor(p_224316_.getHeightAccessorForGeneration());
@@ -327,6 +_,24 @@
         int j = Mth.floorDiv(i, noisesettings.getCellHeight());
         int k = Mth.floorDiv(noisesettings.height(), noisesettings.getCellHeight());
         return k <= 0 ? CompletableFuture.completedFuture(p_224316_) : CompletableFuture.supplyAsync(Util.wrapThreadWithTaskName("wgen_fill_noise", () -> {
+            // Stellar start - Noisium: Optimize noise chunk generation
+            if (USE_NOISIUM_WORLDGEN) {
+                int l = p_224316_.getSectionIndex(k * noisesettings.getCellHeight() - 1 + i);
+                int i1 = p_224316_.getSectionIndex(i);
+
+                var set = p_224316_.getSections();
+                for (int j1 = l; j1 >= i1; --j1) set[j1].acquire();
+
+                ChunkAccess chunkaccess;
+                try {
+                    chunkaccess = this.doFill(p_224313_, p_224315_, p_224314_, p_224316_, j, k);
+                } finally {
+                    for (int j1 = l; j1 >= i1; --j1) set[j1].release();
+                }
+
+                return chunkaccess;
+            }
+            // Stellar end - Noisium: Optimize noise chunk generation
             int l = p_224316_.getSectionIndex(k * noisesettings.getCellHeight() - 1 + i);
             int i1 = p_224316_.getSectionIndex(i);
             Set<LevelChunkSection> set = Sets.newHashSet();
@@ -347,7 +_,7 @@
             }
 
             return chunkaccess;
-        }), Util.backgroundExecutor());
+        }), Runnable::run); // Paper - rewrite chunk system
     }
 
     private ChunkAccess doFill(Blender p_224285_, StructureManager p_224286_, RandomState p_224287_, ChunkAccess p_224288_, int p_224289_, int p_224290_) {
@@ -405,6 +_,24 @@
 
                                 blockstate = this.debugPreliminarySurfaceLevel(noisechunk, l3, l2, k4, blockstate);
                                 if (blockstate != AIR && !SharedConstants.debugVoidTerrain(p_224288_.getPos())) {
+                                    // Stellar start - Noisium: Optimize noise chunk generation
+                                    if (USE_NOISIUM_WORLDGEN) {
+                                        ++levelchunksection.nonEmptyBlockCount;
+                                        if (!blockstate.getFluidState().isEmpty()) ++levelchunksection.tickingFluidCount;
+                                        if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isSpecialCollidingBlock(blockstate))
+                                            ++levelchunksection.specialCollidingBlocks;
+
+                                        if (!blockstate.isRandomlyTicking()) {
+                                            ++levelchunksection.tickingBlockCount;
+                                            levelchunksection.tickingBlocks.remove(i4, i3, l4);
+                                            levelchunksection.tickingBlocks.add(i4, i3, l4, blockstate);
+                                        }
+
+                                        levelchunksection.states.data.storage().set(
+                                            levelchunksection.states.strategy.getIndex(i4, i3, l4),
+                                            levelchunksection.states.data.palette().idFor(blockstate)
+                                        );
+                                    } else
                                     levelchunksection.setBlockState(i4, i3, l4, blockstate, false);
                                     heightmap.update(i4, l2, l4, blockstate);
                                     heightmap1.update(i4, l2, l4, blockstate);
