--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -231,7 +_,7 @@
         }
     }
 
-    interface Condition {
+    public interface Condition { // Paper - public
         boolean test();
     }
 
@@ -257,7 +_,7 @@
         KeyDispatchDataCodec<? extends SurfaceRules.ConditionSource> codec();
     }
 
-    protected static final class Context {
+    public static final class Context { // Paper - public
         private static final int HOW_FAR_BELOW_PRELIMINARY_SURFACE_LEVEL_TO_BUILD_SURFACE = 8;
         private static final int SURFACE_CELL_BITS = 4;
         private static final int SURFACE_CELL_SIZE = 16;
@@ -267,16 +_,16 @@
         final SurfaceRules.Condition steep = new SurfaceRules.Context.SteepMaterialCondition(this);
         final SurfaceRules.Condition hole = new SurfaceRules.Context.HoleCondition(this);
         final SurfaceRules.Condition abovePreliminarySurface = new SurfaceRules.Context.AbovePreliminarySurfaceCondition();
-        final RandomState randomState;
+        public final RandomState randomState; // Paper - public
         final ChunkAccess chunk;
         private final NoiseChunk noiseChunk;
         private final Function<BlockPos, Holder<Biome>> biomeGetter;
-        final WorldGenerationContext context;
+        public final WorldGenerationContext context; // Paper - public
         private long lastPreliminarySurfaceCellOrigin = Long.MAX_VALUE;
         private final int[] preliminarySurfaceCache = new int[4];
         long lastUpdateXZ = -9223372036854775807L;
-        int blockX;
-        int blockZ;
+        public int blockX; // Paper - public
+        public int blockZ; // Paper - public
         int surfaceDepth;
         private long lastSurfaceDepth2Update = this.lastUpdateXZ - 1L;
         private double surfaceSecondary;
@@ -285,7 +_,7 @@
         long lastUpdateY = -9223372036854775807L;
         final BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
         Supplier<Holder<Biome>> biome;
-        int blockY;
+        public int blockY; // Paper - public
         int waterHeight;
         int stoneDepthBelow;
         int stoneDepthAbove;
@@ -403,14 +_,14 @@
                 ChunkAccess chunkaccess = this.context.chunk;
                 int i1 = chunkaccess.getHeight(Heightmap.Types.WORLD_SURFACE_WG, i, k);
                 int j1 = chunkaccess.getHeight(Heightmap.Types.WORLD_SURFACE_WG, i, l);
-                if (j1 >= i1 + 4) {
+                if (Math.abs(j1 - i1) >= 4) { // Stellar - fix MC-258859
                     return true;
                 } else {
                     int k1 = Math.max(i - 1, 0);
                     int l1 = Math.min(i + 1, 15);
                     int i2 = chunkaccess.getHeight(Heightmap.Types.WORLD_SURFACE_WG, k1, j);
                     int j2 = chunkaccess.getHeight(Heightmap.Types.WORLD_SURFACE_WG, l1, j);
-                    return i2 >= j2 + 4;
+                    return Math.abs(i2 - j2) >= 4; // Stellar - fix MC-258859
                 }
             }
         }
@@ -485,7 +_,7 @@
         }
     }
 
-    abstract static class LazyYCondition extends SurfaceRules.LazyCondition {
+    public abstract static class LazyYCondition extends SurfaceRules.LazyCondition { // Paper - public
         protected LazyYCondition(SurfaceRules.Context p_189625_) {
             super(p_189625_);
         }
@@ -570,11 +_,16 @@
         KeyDispatchDataCodec<? extends SurfaceRules.RuleSource> codec();
     }
 
-    static record SequenceRule(List<SurfaceRules.SurfaceRule> rules) implements SurfaceRules.SurfaceRule {
+    static record SequenceRule(List<SurfaceRules.SurfaceRule> rules, SurfaceRules.SurfaceRule[] rulesArray) implements SurfaceRules.SurfaceRule { // Stellar start - Use array for iteration
+
+        SequenceRule(List<SurfaceRules.SurfaceRule> rules) {
+            this(rules, rules.toArray(SurfaceRules.SurfaceRule[]::new));
+        } // Stellar end - Use array for iteration
+
         @Nullable
         @Override
         public BlockState tryApply(int p_189694_, int p_189695_, int p_189696_) {
-            for (SurfaceRules.SurfaceRule surfacerules$surfacerule : this.rules) {
+            for (SurfaceRules.SurfaceRule surfacerules$surfacerule : this.rulesArray) { // Stellar - Use array for iteration
                 BlockState blockstate = surfacerules$surfacerule.tryApply(p_189694_, p_189695_, p_189696_);
                 if (blockstate != null) {
                     return blockstate;
@@ -585,10 +_,14 @@
         }
     }
 
-    static record SequenceRuleSource(List<SurfaceRules.RuleSource> sequence) implements SurfaceRules.RuleSource {
+    static record SequenceRuleSource(List<SurfaceRules.RuleSource> sequence, SurfaceRules.RuleSource[] sequenceArray, boolean isSingleOrNoElement, @Nullable SurfaceRules.RuleSource firstElement) implements SurfaceRules.RuleSource { // Stellar start - Optimize lookup
         static final KeyDispatchDataCodec<SurfaceRules.SequenceRuleSource> CODEC = KeyDispatchDataCodec.of(
             SurfaceRules.RuleSource.CODEC.listOf().xmap(SurfaceRules.SequenceRuleSource::new, SurfaceRules.SequenceRuleSource::sequence).fieldOf("sequence")
         );
+        private static final SurfaceRules.SurfaceRule EMPTY = new SurfaceRules.SequenceRule(List.of());
+        SequenceRuleSource(List<SurfaceRules.RuleSource> sequence) {
+            this(sequence, sequence.toArray(SurfaceRules.RuleSource[]::new), sequence.size() <= 1, sequence.isEmpty() ? null : sequence.getFirst());
+        } // Stellar end - Optimize lookup
 
         @Override
         public KeyDispatchDataCodec<? extends SurfaceRules.RuleSource> codec() {
@@ -596,10 +_,11 @@
         }
 
         public SurfaceRules.SurfaceRule apply(SurfaceRules.Context p_189704_) {
-            if (this.sequence.size() == 1) {
-                return this.sequence.get(0).apply(p_189704_);
+            // Stellar start - Optimize lookup
+            if (this.isSingleOrNoElement) {
+                return this.firstElement != null ? this.firstElement.apply(p_189704_) : EMPTY;
             } else {
-                Builder<SurfaceRules.SurfaceRule> builder = ImmutableList.builder();
+                Builder<SurfaceRules.SurfaceRule> builder = ImmutableList.builderWithExpectedSize(this.sequenceArray.length); // Stellar end - Optimize lookup
 
                 for (SurfaceRules.RuleSource surfacerules$rulesource : this.sequence) {
                     builder.add(surfacerules$rulesource.apply(p_189704_));
