--- a/net/minecraft/client/gui/screens/inventory/AnvilScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/AnvilScreen.java
@@ -130,10 +_,10 @@
     }
 
     @Override
-    public void slotChanged(AbstractContainerMenu p_97882_, int p_97883_, ItemStack p_97884_) {
-        if (p_97883_ == 0) {
-            this.name.setValue(p_97884_.isEmpty() ? "" : p_97884_.getHoverName().getString());
-            this.name.setEditable(!p_97884_.isEmpty());
+    public void slotChanged(AbstractContainerMenu handler, int slotId, ItemStack stack) {
+        if (slotId == 0) {
+            this.name.setValue(stack.isEmpty() ? "" : stack.getHoverName().getString());
+            this.name.setEditable(!stack.isEmpty());
             this.setFocused(this.name);
         }
     }
