--- a/net/minecraft/client/gui/screens/inventory/BeaconScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/BeaconScreen.java
@@ -51,11 +_,11 @@
         this.imageHeight = 219;
         p_97912_.addSlotListener(new ContainerListener() {
             @Override
-            public void slotChanged(AbstractContainerMenu p_97973_, int p_97974_, ItemStack p_97975_) {
+            public void slotChanged(AbstractContainerMenu handler, int slotId, ItemStack stack) {
             }
 
             @Override
-            public void dataChanged(AbstractContainerMenu p_169628_, int p_169629_, int p_169630_) {
+            public void dataChanged(AbstractContainerMenu handler, int property, int value) {
                 BeaconScreen.this.primary = p_97912_.getPrimaryEffect();
                 BeaconScreen.this.secondary = p_97912_.getSecondaryEffect();
             }
