--- a/net/minecraft/client/gui/screens/inventory/ItemCombinerScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/ItemCombinerScreen.java
@@ -55,10 +_,10 @@
     protected abstract void renderErrorIcon(GuiGraphics p_281990_, int p_266822_, int p_267045_);
 
     @Override
-    public void dataChanged(AbstractContainerMenu p_169759_, int p_169760_, int p_169761_) {
+    public void dataChanged(AbstractContainerMenu handler, int property, int value) {
     }
 
     @Override
-    public void slotChanged(AbstractContainerMenu p_98910_, int p_98911_, ItemStack p_98912_) {
+    public void slotChanged(AbstractContainerMenu handler, int slotId, ItemStack stack) {
     }
 }
