--- a/net/minecraft/client/gui/screens/inventory/LecternScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/LecternScreen.java
@@ -17,13 +_,13 @@
     private final LecternMenu menu;
     private final ContainerListener listener = new ContainerListener() {
         @Override
-        public void slotChanged(AbstractContainerMenu p_99054_, int p_99055_, ItemStack p_99056_) {
+        public void slotChanged(AbstractContainerMenu handler, int slotId, ItemStack stack) {
             LecternScreen.this.bookChanged();
         }
 
         @Override
-        public void dataChanged(AbstractContainerMenu p_169772_, int p_169773_, int p_169774_) {
-            if (p_169773_ == 0) {
+        public void dataChanged(AbstractContainerMenu handler, int property, int value) {
+            if (property == 0) {
                 LecternScreen.this.pageChanged();
             }
         }
