--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -29,6 +_,9 @@
 import net.minecraft.world.phys.HitResult;
 
 public class BucketItem extends Item implements DispensibleContainerItem {
+
+    private static @Nullable ItemStack itemLeftInHandAfterPlayerBucketEmptyEvent = null; // Paper - Fix PlayerBucketEmptyEvent result itemstack
+
     public final Fluid content;
 
     public BucketItem(Fluid p_40689_, Item.Properties p_40690_) {
@@ -55,12 +_,22 @@
             } else if (this.content == Fluids.EMPTY) {
                 BlockState blockstate1 = p_40703_.getBlockState(blockpos);
                 if (blockstate1.getBlock() instanceof BucketPickup bucketpickup) {
+                    // CraftBukkit start
+                    ItemStack dummyFluid = bucketpickup.pickupBlock(p_40704_, org.bukkit.craftbukkit.util.DummyGeneratorAccess.INSTANCE, blockpos, blockstate1);
+                    if (dummyFluid.isEmpty()) return InteractionResultHolder.fail(itemstack); // Don't fire event if the bucket won't be filled.
+                    org.bukkit.event.player.PlayerBucketFillEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBucketFillEvent((net.minecraft.server.level.ServerLevel) p_40703_, p_40704_, blockpos, blockpos, blockhitresult.getDirection(), itemstack, dummyFluid.getItem(), p_40705_);
+
+                    if (event.isCancelled()) {
+                        ((ServerPlayer) p_40704_).getBukkitEntity().updateInventory(); // SPIGOT-4541
+                        return InteractionResultHolder.fail(itemstack);
+                    }
+                    // CraftBukkit end
                     ItemStack itemstack3 = bucketpickup.pickupBlock(p_40704_, p_40703_, blockpos, blockstate1);
                     if (!itemstack3.isEmpty()) {
                         p_40704_.awardStat(Stats.ITEM_USED.get(this));
-                        bucketpickup.getPickupSound().ifPresent(p_150709_ -> p_40704_.playSound(p_150709_, 1.0F, 1.0F));
+                        bucketpickup.getPickupSound(blockstate1).ifPresent(p_150709_ -> p_40704_.playSound(p_150709_, 1.0F, 1.0F));
                         p_40703_.gameEvent(p_40704_, GameEvent.FLUID_PICKUP, blockpos);
-                        ItemStack itemstack2 = ItemUtils.createFilledResult(itemstack, p_40704_, itemstack3);
+                        ItemStack itemstack2 = ItemUtils.createFilledResult(itemstack, p_40704_, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItemStack())); // CraftBukkit
                         if (!p_40703_.isClientSide) {
                             CriteriaTriggers.FILLED_BUCKET.trigger((ServerPlayer)p_40704_, itemstack3);
                         }
@@ -72,8 +_,8 @@
                 return InteractionResultHolder.fail(itemstack);
             } else {
                 BlockState blockstate = p_40703_.getBlockState(blockpos);
-                BlockPos blockpos2 = blockstate.getBlock() instanceof LiquidBlockContainer && this.content == Fluids.WATER ? blockpos : blockpos1;
-                if (this.emptyContents(p_40704_, p_40703_, blockpos2, blockhitresult)) {
+                BlockPos blockpos2 = canBlockContainFluid(p_40704_, p_40703_, blockpos, blockstate) ? blockpos : blockpos1;
+                if (this.emptyContents(p_40704_, p_40703_, blockpos2, blockhitresult, blockhitresult.getDirection(), blockpos, itemstack, p_40705_)) {
                     this.checkExtraContent(p_40704_, p_40703_, itemstack, blockpos2);
                     if (p_40704_ instanceof ServerPlayer) {
                         CriteriaTriggers.PLACED_BLOCK.trigger((ServerPlayer)p_40704_, blockpos2, itemstack);
@@ -90,6 +_,13 @@
     }
 
     public static ItemStack getEmptySuccessItem(ItemStack p_40700_, Player p_40701_) {
+        // Paper start - Fix PlayerBucketEmptyEvent result itemstack
+        if (itemLeftInHandAfterPlayerBucketEmptyEvent != null) {
+            ItemStack itemInHand = itemLeftInHandAfterPlayerBucketEmptyEvent;
+            itemLeftInHandAfterPlayerBucketEmptyEvent = null;
+            return itemInHand;
+        }
+        // Paper end - Fix PlayerBucketEmptyEvent result itemstack
         return !p_40701_.hasInfiniteMaterials() ? new ItemStack(Items.BUCKET) : p_40700_;
     }
 
@@ -98,70 +_,100 @@
     }
 
     @Override
+    @Deprecated // Neo: use the ItemStack sensitive version
     public boolean emptyContents(@Nullable Player p_150716_, Level p_150717_, BlockPos p_150718_, @Nullable BlockHitResult p_150719_) {
-        if (!(this.content instanceof FlowingFluid flowingfluid)) {
+        return this.emptyContents(p_150716_, p_150717_, p_150718_, p_150719_, null);
+    }
+    @Override
+    @Deprecated // Stellar - use Direction, BlockPos, ItemStack, InteractionHand sensitive version
+    public boolean emptyContents(@Nullable Player p_150716_, Level p_150717_, BlockPos p_150718_, @Nullable BlockHitResult p_150719_, @Nullable ItemStack container) {
+        return this.emptyContents(p_150716_, p_150717_, p_150718_, p_150719_, null, null, container, InteractionHand.MAIN_HAND);
+    }
+
+    public boolean emptyContents(Player player, Level world, BlockPos blockposition, @Nullable BlockHitResult movingobjectpositionblock, Direction enumdirection, BlockPos clicked, ItemStack itemstack, InteractionHand enumhand) {
+        // CraftBukkit end
+        Fluid fluidtype = this.content;
+
+        if (!(fluidtype instanceof FlowingFluid fluidtypeflowing)) {
             return false;
         } else {
-            Block $$7;
-            boolean $$8;
-            BlockState blockstate;
-            boolean flag2;
-            label82: {
-                blockstate = p_150717_.getBlockState(p_150718_);
-                $$7 = blockstate.getBlock();
-                $$8 = blockstate.canBeReplaced(this.content);
-                label70:
-                if (!blockstate.isAir() && !$$8) {
-                    if ($$7 instanceof LiquidBlockContainer liquidblockcontainer
-                        && liquidblockcontainer.canPlaceLiquid(p_150716_, p_150717_, p_150718_, blockstate, this.content)) {
+            BlockState iblockdata;
+            Block block;
+            boolean flag;
+            LiquidBlockContainer ifluidcontainer;
+            boolean flag1;
+            label70:
+            {
+                iblockdata = world.getBlockState(blockposition);
+                block = iblockdata.getBlock();
+                flag = iblockdata.canBeReplaced(this.content);
+                if (!iblockdata.isAir() && !flag) {
+                    label67:
+                    {
+                        if (block instanceof LiquidBlockContainer) {
+                            ifluidcontainer = (LiquidBlockContainer) block;
+                            if (ifluidcontainer.canPlaceLiquid(player, world, blockposition, iblockdata, this.content)) {
+                                break label67;
+                            }
+                        }
+
+                        flag1 = false;
                         break label70;
                     }
-
-                    flag2 = false;
-                    break label82;
-                }
-
-                flag2 = true;
-            }
-
-            boolean flag1 = flag2;
-            if (!flag1) {
-                return p_150719_ != null && this.emptyContents(p_150716_, p_150717_, p_150719_.getBlockPos().relative(p_150719_.getDirection()), null);
-            } else if (p_150717_.dimensionType().ultraWarm() && this.content.is(FluidTags.WATER)) {
-                int l = p_150718_.getX();
-                int i = p_150718_.getY();
-                int j = p_150718_.getZ();
-                p_150717_.playSound(
-                    p_150716_,
-                    p_150718_,
-                    SoundEvents.FIRE_EXTINGUISH,
-                    SoundSource.BLOCKS,
-                    0.5F,
-                    2.6F + (p_150717_.random.nextFloat() - p_150717_.random.nextFloat()) * 0.8F
-                );
-
-                for (int k = 0; k < 8; k++) {
-                    p_150717_.addParticle(
-                        ParticleTypes.LARGE_SMOKE, (double)l + Math.random(), (double)i + Math.random(), (double)j + Math.random(), 0.0, 0.0, 0.0
-                    );
+                }
+
+                flag1 = true;
+            }
+
+            boolean flag2 = flag1;
+
+            // CraftBukkit start
+            if (flag2 && player != null) {
+                org.bukkit.event.player.PlayerBucketEmptyEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBucketEmptyEvent((net.minecraft.server.level.ServerLevel) world, player, blockposition, clicked, enumdirection, itemstack, enumhand);
+                if (event.isCancelled()) {
+                    // ((ServerPlayer) entityhuman).connection.send(new ClientboundBlockUpdatePacket(world, blockposition)); // SPIGOT-4238: needed when looking through entity // Paper - Don't resend blocks
+                    ((ServerPlayer) player).getBukkitEntity().updateInventory(); // SPIGOT-4541
+                    return false;
+                }
+                itemLeftInHandAfterPlayerBucketEmptyEvent = event.getItemStack() != null ? event.getItemStack().equals(org.bukkit.craftbukkit.inventory.CraftItemStack.asNewCraftStack(Items.BUCKET)) ? null : org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItemStack()) : ItemStack.EMPTY; // Paper - Fix PlayerBucketEmptyEvent result itemstack
+            }
+            // CraftBukkit end
+            java.util.Optional<net.neoforged.neoforge.fluids.FluidStack> containedFluidStack = java.util.Optional.ofNullable(itemstack).flatMap(net.neoforged.neoforge.fluids.FluidUtil::getFluidContained);
+            if (!flag2) {
+                return movingobjectpositionblock != null && this.emptyContents(player, world, movingobjectpositionblock.getBlockPos().relative(movingobjectpositionblock.getDirection()), (BlockHitResult) null, enumdirection, clicked, itemstack, enumhand); // CraftBukkit
+            } else if (containedFluidStack.isPresent() && this.content.getFluidType().isVaporizedOnPlacement(world, blockposition, containedFluidStack.get())) {
+                this.content.getFluidType().onVaporize(player, world, blockposition, containedFluidStack.get());
+                return true;
+            } else if ((world.dimensionType().ultraWarm() || (world.isTheEnd() && !org.purpurmc.purpur.PurpurConfig.allowWaterPlacementInTheEnd)) && this.content.is(FluidTags.WATER)) { // Purpur
+                int i = blockposition.getX();
+                int j = blockposition.getY();
+                int k = blockposition.getZ();
+
+                world.playSound(player, blockposition, SoundEvents.FIRE_EXTINGUISH, SoundSource.BLOCKS, 0.5F, 2.6F + (world.random.nextFloat() - world.random.nextFloat()) * 0.8F);
+
+                for (int l = 0; l < 8; ++l) {
+                    ((net.minecraft.server.level.ServerLevel) world).sendParticles(null, ParticleTypes.LARGE_SMOKE, (double) i + Math.random(), (double) j + Math.random(), (double) k + Math.random(), 1, 0.0D, 0.0D, 0.0D, 0.0D, true); // Purpur
                 }
 
                 return true;
             } else {
-                if ($$7 instanceof LiquidBlockContainer liquidblockcontainer1 && this.content == Fluids.WATER) {
-                    liquidblockcontainer1.placeLiquid(p_150717_, p_150718_, blockstate, flowingfluid.getSource(false));
-                    this.playEmptySound(p_150716_, p_150717_, p_150718_);
-                    return true;
-                }
-
-                if (!p_150717_.isClientSide && $$8 && !blockstate.liquid()) {
-                    p_150717_.destroyBlock(p_150718_, true);
-                }
-
-                if (!p_150717_.setBlock(p_150718_, this.content.defaultFluidState().createLegacyBlock(), 11) && !blockstate.getFluidState().isSource()) {
+                if (block instanceof LiquidBlockContainer) {
+                    ifluidcontainer = (LiquidBlockContainer) block;
+                    if (this.content == Fluids.WATER) {
+                        ifluidcontainer.placeLiquid(world, blockposition, iblockdata, fluidtypeflowing.getSource(false));
+                        this.playEmptySound(player, world, blockposition);
+                        return true;
+                    }
+                }
+
+                if (!world.isClientSide && flag && !iblockdata.liquid()) {
+                    world.destroyBlock(blockposition, true);
+                }
+
+                if (!world.setBlock(blockposition, this.content.defaultFluidState().createLegacyBlock(), 11) && !iblockdata.getFluidState().isSource()) {
                     return false;
                 } else {
-                    this.playEmptySound(p_150716_, p_150717_, p_150718_);
+                    this.playEmptySound(player, world, blockposition);
                     return true;
                 }
             }
@@ -169,8 +_,14 @@
     }
 
     protected void playEmptySound(@Nullable Player p_40696_, LevelAccessor p_40697_, BlockPos p_40698_) {
-        SoundEvent soundevent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
+        SoundEvent soundevent = this.content.getFluidType().getSound(p_40696_, p_40697_, p_40698_, net.neoforged.neoforge.common.SoundActions.BUCKET_EMPTY);
+        if(soundevent == null) soundevent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
         p_40697_.playSound(p_40696_, p_40698_, soundevent, SoundSource.BLOCKS, 1.0F, 1.0F);
         p_40697_.gameEvent(p_40696_, GameEvent.FLUID_PLACE, p_40698_);
+    }
+
+    protected boolean canBlockContainFluid(@Nullable Player player, Level worldIn, BlockPos posIn, BlockState blockstate)
+    {
+        return blockstate.getBlock() instanceof LiquidBlockContainer && ((LiquidBlockContainer)blockstate.getBlock()).canPlaceLiquid(player, worldIn, posIn, blockstate, this.content);
     }
 }
