--- a/net/minecraft/world/item/LeadItem.java
+++ b/net/minecraft/world/item/LeadItem.java
@@ -28,7 +_,7 @@
         if (blockstate.is(BlockTags.FENCES)) {
             Player player = p_42834_.getPlayer();
             if (!level.isClientSide && player != null) {
-                bindPlayerMobs(player, level, blockpos);
+                bindPlayerMobs(player, level, blockpos, p_42834_.getHand()); // CraftBukkit - Pass hand
             }
 
             return InteractionResult.sidedSuccess(level.isClientSide);
@@ -37,16 +_,33 @@
         }
     }
 
-    public static InteractionResult bindPlayerMobs(Player p_42830_, Level p_42831_, BlockPos p_42832_) {
+    public static InteractionResult bindPlayerMobs(Player p_42830_, Level p_42831_, BlockPos p_42832_, net.minecraft.world.InteractionHand enumhand) { // CraftBukkit - Add EnumHand
         LeashFenceKnotEntity leashfenceknotentity = null;
         List<Leashable> list = leashableInArea(p_42831_, p_42832_, p_353025_ -> p_353025_.getLeashHolder() == p_42830_);
-
         for (Leashable leashable : list) {
             if (leashfenceknotentity == null) {
                 leashfenceknotentity = LeashFenceKnotEntity.getOrCreateKnot(p_42831_, p_42832_);
+                // CraftBukkit start - fire HangingPlaceEvent
+
+                org.bukkit.inventory.EquipmentSlot hand = org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(enumhand);
+                org.bukkit.event.hanging.HangingPlaceEvent event = new org.bukkit.event.hanging.HangingPlaceEvent((org.bukkit.entity.Hanging) leashfenceknotentity.getBukkitEntity(), p_42830_ != null ? (org.bukkit.entity.Player) p_42830_.getBukkitEntity() : null, org.bukkit.craftbukkit.block.CraftBlock.at(p_42831_, p_42832_), org.bukkit.block.BlockFace.SELF, hand);
+                p_42831_.getCraftServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    leashfenceknotentity.discard(null); // CraftBukkit - add Bukkit remove cause
+                    return InteractionResult.PASS;
+                }
+                // CraftBukkit end
                 leashfenceknotentity.playPlacementSound();
             }
 
+            // CraftBukkit start
+            if (leashable instanceof Entity leashed) {
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerLeashEntityEvent(leashed, leashfenceknotentity, p_42830_, enumhand).isCancelled()) {
+                    list.remove(leashable);
+                    continue;
+                }
+            }
             leashable.setLeashedTo(leashfenceknotentity, true);
         }
 
@@ -54,9 +_,20 @@
             p_42831_.gameEvent(GameEvent.BLOCK_ATTACH, p_42832_, GameEvent.Context.of(p_42830_));
             return InteractionResult.SUCCESS;
         } else {
+            // CraftBukkit start - remove leash if we do not leash any entity because of the cancelled event
+            if (leashfenceknotentity != null) {
+                leashfenceknotentity.discard(null);
+            }
+            // CraftBukkit end
             return InteractionResult.PASS;
         }
     }
+
+    // CraftBukkit start
+    public static InteractionResult bindPlayerMobs(Player player, Level world, BlockPos pos) {
+        return LeadItem.bindPlayerMobs(player, world, pos, net.minecraft.world.InteractionHand.MAIN_HAND);
+    }
+    // CraftBukkit end
 
     public static List<Leashable> leashableInArea(Level p_353033_, BlockPos p_353031_, Predicate<Leashable> p_353062_) {
         double d0 = 7.0;
