--- a/net/minecraft/world/phys/AABB.java
+++ b/net/minecraft/world/phys/AABB.java
@@ -10,6 +_,7 @@
 
 public class AABB {
     private static final double EPSILON = 1.0E-7;
+    public static final AABB INFINITE = new AABB(Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY);
     public final double minX;
     public final double minY;
     public final double minZ;
@@ -92,11 +_,25 @@
     }
 
     public double min(Direction.Axis p_82341_) {
-        return p_82341_.choose(this.minX, this.minY, this.minZ);
+        // Stellar start - lithium: math.fast_util - Simplify the code to better help the JVM optimize it
+        return switch (p_82341_.ordinal()) {
+            case 0 -> this.minX;
+            case 1 -> this.minY;
+            case 2 -> this.minZ;
+            default -> throw new IncompatibleClassChangeError();
+        };
+        // Stellar end - lithium: math.fast_util - Simplify the code to better help the JVM optimize it
     }
 
     public double max(Direction.Axis p_82375_) {
-        return p_82375_.choose(this.maxX, this.maxY, this.maxZ);
+        // Stellar start - lithium: math.fast_util - Simplify the code to better help the JVM optimize it
+        return switch (p_82375_.ordinal()) {
+            case 0 -> this.maxX;
+            case 1 -> this.maxY;
+            case 2 -> this.maxZ;
+            default -> throw new IncompatibleClassChangeError();
+        };
+        // Stellar end - lithium: math.fast_util - Simplify the code to better help the JVM optimize it
     }
 
     @Override
@@ -340,7 +_,7 @@
     }
 
     @Nullable
-    private static Direction getDirection(
+    public static Direction getDirection( // Paper - optimise collisions - public
         AABB p_82326_, Vec3 p_82327_, double[] p_82328_, @Nullable Direction p_82329_, double p_82330_, double p_82331_, double p_82332_
     ) {
         if (p_82330_ > 1.0E-7) {
@@ -528,5 +_,13 @@
             p_165883_.y + p_165885_ / 2.0,
             p_165883_.z + p_165886_ / 2.0
         );
+    }
+
+    /**
+     * {@return true if this AABB is infinite in all directions}
+     */
+    public boolean isInfinite() {
+        return this == INFINITE || (Double.isInfinite(this.minX) && Double.isInfinite(this.minY) && Double.isInfinite(this.minZ)
+                && Double.isInfinite(this.maxX) && Double.isInfinite(this.maxY) && Double.isInfinite(this.maxZ));
     }
 }
