--- a/net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java
+++ b/net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java
@@ -107,7 +_,7 @@
             this.checkbox = CycleButton.onOffBuilder(p_101104_.get())
                 .displayOnlyValue()
                 .withCustomNarration(p_170219_ -> p_170219_.createDefaultNarrationMessage().append("\n").append(p_101103_))
-                .create(10, 5, 44, 20, p_101101_, (p_170215_, p_170216_) -> p_101104_.set(p_170216_, null));
+                .create(10, 5, 44, 20, p_101101_, (p_170215_, p_170216_) -> p_101104_.set(p_170216_, (net.minecraft.server.level.ServerLevel) null));
             this.children.add(this.checkbox);
         }
 
