--- a/net/minecraft/world/level/levelgen/feature/OreFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/OreFeature.java
@@ -15,6 +_,12 @@
 import net.minecraft.world.level.levelgen.feature.configurations.OreConfiguration;
 
 public class OreFeature extends Feature<OreConfiguration> {
+
+    // Stellar start - Reduce allocations
+    private static final java.util.function.IntFunction<BitSet> bitSetConstructor = BitSet::new;
+    private static final ThreadLocal<it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<BitSet>> BITSETS = ThreadLocal.withInitial(it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap::new);
+    // Stellar end - Reduce allocations
+
     public OreFeature(Codec<OreConfiguration> p_66531_) {
         super(p_66531_);
     }
@@ -69,7 +_,10 @@
         int p_225185_
     ) {
         int i = 0;
-        BitSet bitset = new BitSet(p_225184_ * p_225185_ * p_225184_);
+        // Stellar start - Reduce allocations
+        BitSet bitset = BITSETS.get().computeIfAbsent(p_225184_ * p_225185_ * p_225184_, bitSetConstructor);
+        bitset.clear();
+        // Stellar end - Reduce allocations
         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
         int j = p_225174_.size;
         double[] adouble = new double[j * 4];
@@ -137,9 +_,9 @@
                                                 if (p_225172_.ensureCanWrite(blockpos$mutableblockpos)) {
                                                     LevelChunkSection levelchunksection = bulksectionaccess.getSection(blockpos$mutableblockpos);
                                                     if (levelchunksection != null) {
-                                                        int i3 = SectionPos.sectionRelative(i2);
-                                                        int j3 = SectionPos.sectionRelative(j2);
-                                                        int k3 = SectionPos.sectionRelative(k2);
+                                                        int i3 = i2 & 15;
+                                                        int j3 = j2 & 15;
+                                                        int k3 = k2 & 15;
                                                         BlockState blockstate = levelchunksection.getBlockState(i3, j3, k3);
 
                                                         for (OreConfiguration.TargetBlockState oreconfiguration$targetblockstate : p_225174_.targetStates) {
