--- a/net/minecraft/world/level/block/state/StateHolder.java
+++ b/net/minecraft/world/level/block/state/StateHolder.java
@@ -149,6 +_,10 @@
             }
 
             this.neighbours = (Table<Property<?>, Comparable<?>, S>)(table.isEmpty() ? table : ArrayTable.create(table));
+
+            if (this.owner instanceof net.minecraft.world.level.block.Block || this.owner instanceof net.minecraft.world.level.material.Fluid) {
+                this.neighbours = new net.gensokyoreimagined.nitori.common.state.FastImmutableTable<Property<?>, Comparable<?>, S>(this.neighbours, net.gensokyoreimagined.nitori.common.state.StatePropertyTableCache.getTableCache(this.owner)); // Stellar - Nitori
+            }
         }
     }
 
