--- a/net/minecraft/world/ticks/LevelTicks.java
+++ b/net/minecraft/world/ticks/LevelTicks.java
@@ -193,6 +_,7 @@
     }
 
     private void cleanupAfterTick() {
+        if (false) // Stellar - Leaf: Avoid useless deque clear on LevelTicks#cleanupAfterTick - This method runs after toRunThisTick is polled so this is always empty
         this.toRunThisTick.clear();
         this.containersToTick.clear();
         this.alreadyRunThisTick.clear();
