--- a/net/minecraft/core/cauldron/CauldronInteraction.java
+++ b/net/minecraft/core/cauldron/CauldronInteraction.java
@@ -37,19 +_,20 @@
     CauldronInteraction.InteractionMap WATER = newInteractionMap("water");
     CauldronInteraction.InteractionMap LAVA = newInteractionMap("lava");
     CauldronInteraction.InteractionMap POWDER_SNOW = newInteractionMap("powder_snow");
-    CauldronInteraction FILL_WATER = (p_315877_, p_315878_, p_315879_, p_315880_, p_315881_, p_315882_) -> emptyBucket(
+    CauldronInteraction FILL_WATER = (p_315877_, p_315878_, p_315879_, p_315880_, p_315881_, p_315882_, hitDirection) -> emptyBucket( // Paper - add hitDirection
             p_315878_,
             p_315879_,
             p_315880_,
             p_315881_,
             p_315882_,
             Blocks.WATER_CAULDRON.defaultBlockState().setValue(LayeredCauldronBlock.LEVEL, Integer.valueOf(3)),
-            SoundEvents.BUCKET_EMPTY
-        );
-    CauldronInteraction FILL_LAVA = (p_315853_, p_315854_, p_315855_, p_315856_, p_315857_, p_315858_) -> emptyBucket(
-            p_315854_, p_315855_, p_315856_, p_315857_, p_315858_, Blocks.LAVA_CAULDRON.defaultBlockState(), SoundEvents.BUCKET_EMPTY_LAVA
-        );
-    CauldronInteraction FILL_POWDER_SNOW = (p_315871_, p_315872_, p_315873_, p_315874_, p_315875_, p_315876_) -> emptyBucket(
+            SoundEvents.BUCKET_EMPTY,
+            hitDirection // Paper - add hitDirection
+        );
+    CauldronInteraction FILL_LAVA = (p_315853_, p_315854_, p_315855_, p_315856_, p_315857_, p_315858_, hitDirection) -> emptyBucket( // Paper - add hitDirection
+            p_315854_, p_315855_, p_315856_, p_315857_, p_315858_, Blocks.LAVA_CAULDRON.defaultBlockState(), SoundEvents.BUCKET_EMPTY_LAVA, hitDirection // Paper - add hitDirection
+        );
+    CauldronInteraction FILL_POWDER_SNOW = (p_315871_, p_315872_, p_315873_, p_315874_, p_315875_, p_315876_, hitDirection) -> emptyBucket( // Paper - add hitDirection
             p_315872_,
             p_315873_,
             p_315874_,
@@ -58,47 +_,62 @@
             Blocks.POWDER_SNOW_CAULDRON.defaultBlockState().setValue(LayeredCauldronBlock.LEVEL, Integer.valueOf(3)),
             SoundEvents.BUCKET_EMPTY_POWDER_SNOW
         );
-    CauldronInteraction SHULKER_BOX = (p_346995_, p_346996_, p_346997_, p_346998_, p_346999_, p_347000_) -> {
+    CauldronInteraction SHULKER_BOX = (p_346995_, p_346996_, p_346997_, p_346998_, p_346999_, p_347000_, hitDirection) -> { // Paper - add hitDirection
         Block block = Block.byItem(p_347000_.getItem());
         if (!(block instanceof ShulkerBoxBlock)) {
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else {
             if (!p_346996_.isClientSide) {
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.lowerFillLevel(p_346995_, p_346996_, p_346997_, p_346998_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.SHULKER_WASH)) {
+                    return ItemInteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 ItemStack itemstack = p_347000_.transmuteCopy(Blocks.SHULKER_BOX, 1);
                 p_346998_.setItemInHand(p_346999_, ItemUtils.createFilledResult(p_347000_, p_346998_, itemstack, false));
                 p_346998_.awardStat(Stats.CLEAN_SHULKER_BOX);
-                LayeredCauldronBlock.lowerFillLevel(p_346995_, p_346996_, p_346997_);
+                // LayeredCauldronBlock.lowerFillLevel(p_346995_, p_346996_, p_346997_); // CraftBukkit
             }
 
             return ItemInteractionResult.sidedSuccess(p_346996_.isClientSide);
         }
     };
-    CauldronInteraction BANNER = (p_347001_, p_347002_, p_347003_, p_347004_, p_347005_, p_347006_) -> {
+    CauldronInteraction BANNER = (p_347001_, p_347002_, p_347003_, p_347004_, p_347005_, p_347006_, hitDirection) -> { // Paper - add hitDirection
         BannerPatternLayers bannerpatternlayers = p_347006_.getOrDefault(DataComponents.BANNER_PATTERNS, BannerPatternLayers.EMPTY);
         if (bannerpatternlayers.layers().isEmpty()) {
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else {
             if (!p_347002_.isClientSide) {
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.lowerFillLevel(p_347001_, p_347002_, p_347003_, p_347004_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.BANNER_WASH)) {
+                    return ItemInteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 ItemStack itemstack = p_347006_.copyWithCount(1);
                 itemstack.set(DataComponents.BANNER_PATTERNS, bannerpatternlayers.removeLast());
                 p_347004_.setItemInHand(p_347005_, ItemUtils.createFilledResult(p_347006_, p_347004_, itemstack, false));
                 p_347004_.awardStat(Stats.CLEAN_BANNER);
-                LayeredCauldronBlock.lowerFillLevel(p_347001_, p_347002_, p_347003_);
+                // LayeredCauldronBlock.lowerFillLevel(p_347001_, p_347002_, p_347003_); // CraftBukkit
             }
 
             return ItemInteractionResult.sidedSuccess(p_347002_.isClientSide);
         }
     };
-    CauldronInteraction DYED_ITEM = (p_329837_, p_329838_, p_329839_, p_329840_, p_329841_, p_329842_) -> {
+    CauldronInteraction DYED_ITEM = (p_329837_, p_329838_, p_329839_, p_329840_, p_329841_, p_329842_, hitDirection) -> { // Paper - add hitDirection
         if (!p_329842_.is(ItemTags.DYEABLE)) {
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else if (!p_329842_.has(DataComponents.DYED_COLOR)) {
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else {
             if (!p_329838_.isClientSide) {
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.lowerFillLevel(p_329837_, p_329838_, p_329839_, p_329840_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.ARMOR_WASH)) {
+                    return ItemInteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 p_329842_.remove(DataComponents.DYED_COLOR);
                 p_329840_.awardStat(Stats.CLEAN_ARMOR);
-                LayeredCauldronBlock.lowerFillLevel(p_329837_, p_329838_, p_329839_);
+                // LayeredCauldronBlock.lowerFillLevel(p_329837_, p_329838_, p_329839_); // CraftBukkit
             }
 
             return ItemInteractionResult.sidedSuccess(p_329838_.isClientSide);
@@ -108,27 +_,32 @@
     static CauldronInteraction.InteractionMap newInteractionMap(String p_304841_) {
         Object2ObjectOpenHashMap<Item, CauldronInteraction> object2objectopenhashmap = new Object2ObjectOpenHashMap<>();
         object2objectopenhashmap.defaultReturnValue(
-            (p_315883_, p_315884_, p_315885_, p_315886_, p_315887_, p_315888_) -> ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION
+            (p_315883_, p_315884_, p_315885_, p_315886_, p_315887_, p_315888_, hitDirection) -> ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION // Paper - add hitDirection
         );
         CauldronInteraction.InteractionMap cauldroninteraction$interactionmap = new CauldronInteraction.InteractionMap(p_304841_, object2objectopenhashmap);
         INTERACTIONS.put(p_304841_, cauldroninteraction$interactionmap);
         return cauldroninteraction$interactionmap;
     }
 
-    ItemInteractionResult interact(BlockState p_175711_, Level p_175712_, BlockPos p_175713_, Player p_175714_, InteractionHand p_175715_, ItemStack p_175716_);
+    ItemInteractionResult interact(BlockState p_175711_, Level p_175712_, BlockPos p_175713_, Player p_175714_, InteractionHand p_175715_, ItemStack p_175716_, final net.minecraft.core.Direction hitDirection); // Paper - add hitDirection
 
     static void bootStrap() {
         Map<Item, CauldronInteraction> map = EMPTY.map();
         addDefaultInteractions(map);
-        map.put(Items.POTION, (p_329825_, p_329826_, p_329827_, p_329828_, p_329829_, p_329830_) -> {
+        map.put(Items.POTION, (p_329825_, p_329826_, p_329827_, p_329828_, p_329829_, p_329830_, hitDirection) -> { // Paper - add hitDirection
             PotionContents potioncontents = p_329830_.get(DataComponents.POTION_CONTENTS);
             if (potioncontents != null && potioncontents.is(Potions.WATER)) {
                 if (!p_329826_.isClientSide) {
+                    // CraftBukkit start
+                    if (!LayeredCauldronBlock.changeLevel(p_329825_, p_329826_, p_329827_, Blocks.WATER_CAULDRON.defaultBlockState(), p_329828_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                        return ItemInteractionResult.SUCCESS;
+                    }
+                    // CraftBukkit end
                     Item item = p_329830_.getItem();
                     p_329828_.setItemInHand(p_329829_, ItemUtils.createFilledResult(p_329830_, p_329828_, new ItemStack(Items.GLASS_BOTTLE)));
                     p_329828_.awardStat(Stats.USE_CAULDRON);
                     p_329828_.awardStat(Stats.ITEM_USED.get(item));
-                    p_329826_.setBlockAndUpdate(p_329827_, Blocks.WATER_CAULDRON.defaultBlockState());
+                    // p_329826_.setBlockAndUpdate(p_329827_, Blocks.WATER_CAULDRON.defaultBlockState()); // CraftBukkit
                     p_329826_.playSound(null, p_329827_, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                     p_329826_.gameEvent(null, GameEvent.FLUID_PLACE, p_329827_);
                 }
@@ -142,7 +_,7 @@
         addDefaultInteractions(map1);
         map1.put(
             Items.BUCKET,
-            (p_315865_, p_315866_, p_315867_, p_315868_, p_315869_, p_315870_) -> fillBucket(
+            (p_315865_, p_315866_, p_315867_, p_315868_, p_315869_, p_315870_, hitDirection) -> fillBucket( // Paper - add hitDirection
                     p_315865_,
                     p_315866_,
                     p_315867_,
@@ -151,20 +_,26 @@
                     p_315870_,
                     new ItemStack(Items.WATER_BUCKET),
                     p_175660_ -> p_175660_.getValue(LayeredCauldronBlock.LEVEL) == 3,
-                    SoundEvents.BUCKET_FILL
+                    SoundEvents.BUCKET_FILL,
+                    hitDirection // Paper - add hitDirection
                 )
         );
         map1.put(
             Items.GLASS_BOTTLE,
-            (p_329819_, p_329820_, p_329821_, p_329822_, p_329823_, p_329824_) -> {
+            (p_329819_, p_329820_, p_329821_, p_329822_, p_329823_, p_329824_, hitDirection) -> { // Paper - add hitDirection
                 if (!p_329820_.isClientSide) {
+                    // CraftBukkit start
+                    if (!LayeredCauldronBlock.lowerFillLevel(p_329819_, p_329820_, p_329821_, p_329822_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+                        return ItemInteractionResult.SUCCESS;
+                    }
+                    // CraftBukkit end
                     Item item = p_329824_.getItem();
                     p_329822_.setItemInHand(
                         p_329823_, ItemUtils.createFilledResult(p_329824_, p_329822_, PotionContents.createItemStack(Items.POTION, Potions.WATER))
                     );
                     p_329822_.awardStat(Stats.USE_CAULDRON);
                     p_329822_.awardStat(Stats.ITEM_USED.get(item));
-                    LayeredCauldronBlock.lowerFillLevel(p_329819_, p_329820_, p_329821_);
+                    // LayeredCauldronBlock.lowerFillLevel(p_329819_, p_329820_, p_329821_);
                     p_329820_.playSound(null, p_329821_, SoundEvents.BOTTLE_FILL, SoundSource.BLOCKS, 1.0F, 1.0F);
                     p_329820_.gameEvent(null, GameEvent.FLUID_PICKUP, p_329821_);
                 }
@@ -172,17 +_,22 @@
                 return ItemInteractionResult.sidedSuccess(p_329820_.isClientSide);
             }
         );
-        map1.put(Items.POTION, (p_175704_, p_175705_, p_175706_, p_175707_, p_175708_, p_175709_) -> {
+        map1.put(Items.POTION, (p_175704_, p_175705_, p_175706_, p_175707_, p_175708_, p_175709_, hitDirection) -> { // Paper - add hitDirection
             if (p_175704_.getValue(LayeredCauldronBlock.LEVEL) == 3) {
                 return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
             } else {
                 PotionContents potioncontents = p_175709_.get(DataComponents.POTION_CONTENTS);
                 if (potioncontents != null && potioncontents.is(Potions.WATER)) {
                     if (!p_175705_.isClientSide) {
+                        // CraftBukkit start
+                        if (!LayeredCauldronBlock.changeLevel(p_175704_, p_175705_, p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL), p_175707_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY, false)) { // Paper - Call org.bukkit.event.block.CauldronLevelChangeEvent
+                            return ItemInteractionResult.SUCCESS;
+                        }
+                        // CraftBukkit end
                         p_175707_.setItemInHand(p_175708_, ItemUtils.createFilledResult(p_175709_, p_175707_, new ItemStack(Items.GLASS_BOTTLE)));
                         p_175707_.awardStat(Stats.USE_CAULDRON);
                         p_175707_.awardStat(Stats.ITEM_USED.get(p_175709_.getItem()));
-                        p_175705_.setBlockAndUpdate(p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL));
+                        // p_175705_.setBlockAndUpdate(p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL)); // CraftBukkit
                         p_175705_.playSound(null, p_175706_, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                         p_175705_.gameEvent(null, GameEvent.FLUID_PLACE, p_175706_);
                     }
@@ -234,7 +_,7 @@
         Map<Item, CauldronInteraction> map2 = LAVA.map();
         map2.put(
             Items.BUCKET,
-            (p_315889_, p_315890_, p_315891_, p_315892_, p_315893_, p_315894_) -> fillBucket(
+            (p_315889_, p_315890_, p_315891_, p_315892_, p_315893_, p_315894_, hitDirection) -> fillBucket( // Paper - add hitDirection
                     p_315889_,
                     p_315890_,
                     p_315891_,
@@ -243,14 +_,15 @@
                     p_315894_,
                     new ItemStack(Items.LAVA_BUCKET),
                     p_175651_ -> true,
-                    SoundEvents.BUCKET_FILL_LAVA
+                    SoundEvents.BUCKET_FILL_LAVA,
+                    hitDirection // Paper - add hitDirection
                 )
         );
         addDefaultInteractions(map2);
         Map<Item, CauldronInteraction> map3 = POWDER_SNOW.map();
         map3.put(
             Items.BUCKET,
-            (p_315859_, p_315860_, p_315861_, p_315862_, p_315863_, p_315864_) -> fillBucket(
+            (p_315859_, p_315860_, p_315861_, p_315862_, p_315863_, p_315864_, hitDirection) -> fillBucket( // Paper - add hitDirection
                     p_315859_,
                     p_315860_,
                     p_315861_,
@@ -282,15 +_,45 @@
         Predicate<BlockState> p_175643_,
         SoundEvent p_175644_
     ) {
+        // Paper start - add hitDirection
+        return fillBucket(p_175636_, p_175637_, p_175638_, p_175639_, p_175640_, p_175641_, p_175642_, p_175643_, p_175644_, null); // Paper - add hitDirection
+    }
+    static ItemInteractionResult fillBucket(
+        BlockState p_175636_,
+        Level p_175637_,
+        BlockPos p_175638_,
+        Player p_175639_,
+        InteractionHand p_175640_,
+        ItemStack p_175641_,
+        ItemStack p_175642_,
+        Predicate<BlockState> p_175643_,
+        SoundEvent p_175644_,
+        @javax.annotation.Nullable net.minecraft.core.Direction hitDirection
+    ) {
+        // Paper end - add hitDirection
         if (!p_175643_.test(p_175636_)) {
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else {
             if (!p_175637_.isClientSide) {
+                // Paper start - fire PlayerBucketFillEvent
+                if (hitDirection != null) {
+                    org.bukkit.event.player.PlayerBucketEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBucketFillEvent((net.minecraft.server.level.ServerLevel) p_175637_, p_175639_, p_175638_, p_175638_, hitDirection, p_175641_, p_175642_.getItem(), p_175640_);
+                    if (event.isCancelled()) {
+                        return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
+                    }
+                    p_175642_ = event.getItemStack() != null ? org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItemStack()) : ItemStack.EMPTY;
+                }
+                // Paper end - fire PlayerBucketFillEvent
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.changeLevel(p_175636_, p_175637_, p_175638_, Blocks.CAULDRON.defaultBlockState(), p_175639_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.BUCKET_FILL)) {
+                    return ItemInteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 Item item = p_175641_.getItem();
                 p_175639_.setItemInHand(p_175640_, ItemUtils.createFilledResult(p_175641_, p_175639_, p_175642_));
                 p_175639_.awardStat(Stats.USE_CAULDRON);
                 p_175639_.awardStat(Stats.ITEM_USED.get(item));
-                p_175637_.setBlockAndUpdate(p_175638_, Blocks.CAULDRON.defaultBlockState());
+                // p_175637_.setBlockAndUpdate(p_175638_, Blocks.CAULDRON.defaultBlockState()); // CraftBukkit
                 p_175637_.playSound(null, p_175638_, p_175644_, SoundSource.BLOCKS, 1.0F, 1.0F);
                 p_175637_.gameEvent(null, GameEvent.FLUID_PICKUP, p_175638_);
             }
@@ -302,12 +_,34 @@
     static ItemInteractionResult emptyBucket(
         Level p_175619_, BlockPos p_175620_, Player p_175621_, InteractionHand p_175622_, ItemStack p_175623_, BlockState p_175624_, SoundEvent p_175625_
     ) {
+        // Paper start - add hitDirection
+        return emptyBucket(p_175619_, p_175620_, p_175621_, p_175622_, p_175623_, p_175624_, p_175625_, null);
+    }
+    static ItemInteractionResult emptyBucket(
+        Level p_175619_, BlockPos p_175620_, Player p_175621_, InteractionHand p_175622_, ItemStack p_175623_, BlockState p_175624_, SoundEvent p_175625_, @javax.annotation.Nullable net.minecraft.core.Direction hitDirection
+    ) {
+        // Paper end - add hitDirection
         if (!p_175619_.isClientSide) {
+            // Paper start - fire PlayerBucketEmptyEvent
+            ItemStack output = new ItemStack(Items.BUCKET);
+            if (hitDirection != null) {
+                org.bukkit.event.player.PlayerBucketEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBucketEmptyEvent((net.minecraft.server.level.ServerLevel) p_175619_, p_175621_, p_175620_, p_175620_, hitDirection, p_175623_, p_175622_);
+                if (event.isCancelled()) {
+                    return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
+                }
+                output = event.getItemStack() != null ? org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItemStack()) : ItemStack.EMPTY;
+            }
+            // Paper end - fire PlayerBucketEmptyEvent
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.changeLevel(p_175624_, p_175619_, p_175620_, p_175624_, p_175621_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.BUCKET_EMPTY, false)) { // Paper - Call CauldronLevelChangeEvent
+                return ItemInteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             Item item = p_175623_.getItem();
-            p_175621_.setItemInHand(p_175622_, ItemUtils.createFilledResult(p_175623_, p_175621_, new ItemStack(Items.BUCKET)));
+            p_175621_.setItemInHand(p_175622_, ItemUtils.createFilledResult(p_175623_, p_175621_, output)); // Paper
             p_175621_.awardStat(Stats.FILL_CAULDRON);
             p_175621_.awardStat(Stats.ITEM_USED.get(item));
-            p_175619_.setBlockAndUpdate(p_175620_, p_175624_);
+            // p_175619_.setBlockAndUpdate(p_175620_, p_175624_); // CraftBukkit
             p_175619_.playSound(null, p_175620_, p_175625_, SoundSource.BLOCKS, 1.0F, 1.0F);
             p_175619_.gameEvent(null, GameEvent.FLUID_PLACE, p_175620_);
         }
