--- a/net/minecraft/world/entity/monster/Guardian.java
+++ b/net/minecraft/world/entity/monster/Guardian.java
@@ -58,6 +_,7 @@
     private boolean clientSideTouchedGround;
     @Nullable
     public RandomStrollGoal randomStrollGoal;
+    public Guardian.GuardianAttackGoal guardianAttackGoal; // CraftBukkit - add field
 
     public Guardian(EntityType<? extends Guardian> p_32810_, Level p_32811_) {
         super(p_32810_, p_32811_);
@@ -72,7 +_,7 @@
     protected void registerGoals() {
         MoveTowardsRestrictionGoal movetowardsrestrictiongoal = new MoveTowardsRestrictionGoal(this, 1.0);
         this.randomStrollGoal = new RandomStrollGoal(this, 1.0, 80);
-        this.goalSelector.addGoal(4, new Guardian.GuardianAttackGoal(this));
+        this.goalSelector.addGoal(4,  guardianAttackGoal = new Guardian.GuardianAttackGoal(this));
         this.goalSelector.addGoal(5, movetowardsrestrictiongoal);
         this.goalSelector.addGoal(7, this.randomStrollGoal);
         this.goalSelector.addGoal(8, new LookAtPlayerGoal(this, Player.class, 8.0F));
@@ -169,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return this.isInWaterOrBubble() ? SoundEvents.GUARDIAN_DEATH : SoundEvents.GUARDIAN_DEATH_LAND;
     }
 
@@ -318,14 +_,14 @@
     }
 
     @Override
-    public boolean hurt(DamageSource p_32820_, float p_32821_) {
+    public boolean hurt(DamageSource source, float amount) {
         if (this.level().isClientSide) {
             return false;
         } else {
             if (!this.isMoving()
-                && !p_32820_.is(DamageTypeTags.AVOIDS_GUARDIAN_THORNS)
-                && !p_32820_.is(DamageTypes.THORNS)
-                && p_32820_.getDirectEntity() instanceof LivingEntity livingentity) {
+                && !source.is(DamageTypeTags.AVOIDS_GUARDIAN_THORNS)
+                && !source.is(DamageTypes.THORNS)
+                && source.getDirectEntity() instanceof LivingEntity livingentity) {
                 livingentity.hurt(this.damageSources().thorns(this), 2.0F);
             }
 
@@ -333,7 +_,7 @@
                 this.randomStrollGoal.trigger();
             }
 
-            return super.hurt(p_32820_, p_32821_);
+            return super.hurt(source, amount);
         }
     }
 
