--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -94,6 +_,34 @@
         this.setPathfindingMalus(PathType.DAMAGE_FIRE, 0.0F);
     }
 
+    // Purpur start
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.striderRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.striderRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.striderControllable;
+    }
+    // Purpur end
+
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.striderMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.striderScale);
+    }
+
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level().purpurConfig.striderBreedingTicks;
+    }
+
     public static boolean checkStriderSpawnRules(
         EntityType<Strider> p_219129_, LevelAccessor p_219130_, MobSpawnType p_219131_, BlockPos p_219132_, RandomSource p_219133_
     ) {
@@ -321,7 +_,14 @@
             }
 
             boolean flag1 = flag2;
-            this.setSuffocating(!flag || flag1);
+            // CraftBukkit start
+            boolean suffocating = !flag || flag1;
+            if (suffocating ^ this.isSuffocating()) {
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callStriderTemperatureChangeEvent(this, suffocating)) {
+                    this.setSuffocating(suffocating);
+                }
+            }
+            // CraftBukkit end
         }
 
         super.tick();
@@ -365,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.STRIDER_DEATH;
     }
 
@@ -529,10 +_,23 @@
             super(p_33969_, p_33970_);
         }
 
+        // Stellar start - Process pathfinding asynchronously
+        private static final dev.kaiijumc.kaiiju.path.NodeEvaluatorGenerator GENERATOR =
+            (dev.kaiijumc.kaiiju.path.NodeEvaluatorFeatures features) -> {
+                WalkNodeEvaluator evaluator = new WalkNodeEvaluator();
+                evaluator.setCanPassDoors(features.canPassDoors());
+                evaluator.setCanOpenDoors(features.canOpenDoors());
+                evaluator.setCanWalkOverFences(features.canWalkOverFences());
+                evaluator.setCanFloat(features.canFloat());
+                return evaluator;
+            };
+        // Stellar end - Process pathfinding asynchronously
+
         @Override
         protected PathFinder createPathFinder(int p_33972_) {
             this.nodeEvaluator = new WalkNodeEvaluator();
             this.nodeEvaluator.setCanPassDoors(true);
+            if (dev.instellar.stellar.configuration.GlobalConfiguration.get().entities.asyncPathProcess.enabled) return new PathFinder(this.nodeEvaluator, p_33972_, GENERATOR); // Stellar - Process pathfinding asynchronously
             return new PathFinder(this.nodeEvaluator, p_33972_);
         }
 
