--- a/net/minecraft/world/entity/monster/Zoglin.java
+++ b/net/minecraft/world/entity/monster/Zoglin.java
@@ -194,12 +_,12 @@
     }
 
     @Override
-    public boolean hurt(DamageSource p_34214_, float p_34215_) {
-        boolean flag = super.hurt(p_34214_, p_34215_);
+    public boolean hurt(DamageSource source, float amount) {
+        boolean flag = super.hurt(source, amount);
         if (this.level().isClientSide) {
             return false;
-        } else if (flag && p_34214_.getEntity() instanceof LivingEntity) {
-            LivingEntity livingentity = (LivingEntity)p_34214_.getEntity();
+        } else if (flag && source.getEntity() instanceof LivingEntity) {
+            LivingEntity livingentity = (LivingEntity)source.getEntity();
             if (this.canAttack(livingentity) && !BehaviorUtils.isOtherTargetMuchFurtherAwayThanCurrentAttackTarget(this, livingentity, 4.0)) {
                 this.setAttackTarget(livingentity);
             }
@@ -291,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.ZOGLIN_DEATH;
     }
 
