--- a/net/minecraft/server/level/DemoMode.java
+++ b/net/minecraft/server/level/DemoMode.java
@@ -68,11 +_,11 @@
     }
 
     @Override
-    public void handleBlockBreakAction(BlockPos p_214976_, ServerboundPlayerActionPacket.Action p_214977_, Direction p_214978_, int p_214979_, int p_214980_) {
+    public void handleBlockBreakAction(BlockPos pos, ServerboundPlayerActionPacket.Action action, Direction direction, int worldHeight, int sequence) {
         if (this.demoHasEnded) {
             this.outputDemoReminder();
         } else {
-            super.handleBlockBreakAction(p_214976_, p_214977_, p_214978_, p_214979_, p_214980_);
+            super.handleBlockBreakAction(pos, action, direction, worldHeight, sequence);
         }
     }
 
