--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -13,7 +_,13 @@
 
 public class DerivedLevelData implements ServerLevelData {
     private final WorldData worldData;
-    private final ServerLevelData wrapped;
+    public final ServerLevelData wrapped;
+    // Youer start
+    private net.minecraft.resources.ResourceKey<net.minecraft.world.level.dimension.LevelStem> typeKey;
+    public void setTypeKey(net.minecraft.resources.ResourceKey<net.minecraft.world.level.dimension.LevelStem> typeKey) {
+        this.typeKey = typeKey;
+    }
+    // Youer end
 
     public DerivedLevelData(WorldData p_78079_, ServerLevelData p_78080_) {
         this.worldData = p_78079_;
@@ -42,7 +_,22 @@
 
     @Override
     public String getLevelName() {
-        return this.worldData.getLevelName();
+        if (typeKey == null || typeKey == net.minecraft.world.level.dimension.LevelStem.OVERWORLD) { // Youer
+        return this.wrapped.getLevelName();
+        // Youer start
+        } else {
+            String worldName = this.wrapped.getLevelName() + "/";
+            String suffix;
+            if (typeKey == net.minecraft.world.level.dimension.LevelStem.END) {
+                suffix = "DIM1";
+            } else if (typeKey == net.minecraft.world.level.dimension.LevelStem.NETHER) {
+                suffix = "DIM-1";
+            } else {
+                suffix = typeKey.location().getNamespace() + "/" + typeKey.location().getPath();
+            }
+            return worldName + suffix;
+        }
+        // Youer end
     }
 
     @Override
@@ -95,6 +_,12 @@
     public void setThundering(boolean p_78100_) {
     }
 
+    // Stellar start
+    @Override
+    public void setThundering(boolean thundering, org.bukkit.event.weather.ThunderChangeEvent.Cause cause) {
+    }
+    // Stellar end
+
     @Override
     public void setThunderTime(int p_78118_) {
     }
@@ -104,6 +_,10 @@
     }
 
     @Override
+    public void setRaining(boolean raining, org.bukkit.event.weather.WeatherChangeEvent.Cause cause) {
+    }
+
+    @Override
     public void setRainTime(int p_78121_) {
     }
 
@@ -190,5 +_,27 @@
     public void fillCrashReportCategory(CrashReportCategory p_164852_, LevelHeightAccessor p_164853_) {
         p_164852_.setDetail("Derived", true);
         this.wrapped.fillCrashReportCategory(p_164852_, p_164853_);
+    }
+
+    // Neo: Variable day time code
+
+    @Override
+    public float getDayTimeFraction() {
+        return wrapped.getDayTimeFraction();
+    }
+
+    @Override
+    public float getDayTimePerTick() {
+        return wrapped.getDayTimePerTick();
+    }
+
+    @Override
+    public void setDayTimeFraction(float dayTimeFraction) {
+        wrapped.setDayTimeFraction(dayTimeFraction);
+    }
+
+    @Override
+    public void setDayTimePerTick(float dayTimePerTick) {
+        wrapped.setDayTimePerTick(dayTimePerTick);
     }
 }
