--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -261,6 +_,11 @@
         }
 
         @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+            return null;
+        }
+
+        @Override
         public void close() {
             IOUtils.closeQuietly((Writer)this.output);
         }
