--- a/net/minecraft/server/commands/OpCommand.java
+++ b/net/minecraft/server/commands/OpCommand.java
@@ -49,7 +_,7 @@
             if (!playerlist.isOp(gameprofile)) {
                 playerlist.op(gameprofile);
                 i++;
-                p_138089_.sendSuccess(() -> Component.translatable("commands.op.success", p_138090_.iterator().next().getName()), true);
+                p_138089_.sendSuccess(() -> Component.translatable("commands.op.success", gameprofile.getName()), true); // Paper - fixes MC-253721
             }
         }
 
