--- /dev/null
+++ b/net/minecraft/server/commands/SpreadFakePlayersCommand.java
@@ -1,0 +_,134 @@
+package net.minecraft.server.commands;
+
+import com.mojang.authlib.GameProfile;
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.arguments.BoolArgumentType;
+import com.mojang.brigadier.arguments.FloatArgumentType;
+import com.mojang.brigadier.arguments.IntegerArgumentType;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import io.netty.channel.embedded.EmbeddedChannel;
+import it.unimi.dsi.fastutil.objects.ReferenceArraySet;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.Commands;
+import net.minecraft.commands.arguments.DimensionArgument;
+import net.minecraft.commands.arguments.EntityArgument;
+import net.minecraft.commands.arguments.coordinates.Vec2Argument;
+import net.minecraft.network.Connection;
+import net.minecraft.network.protocol.PacketFlow;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.network.CommonListenerCookie;
+import net.neoforged.neoforge.common.util.FakePlayerFactory;
+
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Consumer;
+
+import static net.minecraft.server.commands.SpreadPlayersCommand.spreadPlayers;
+
+public final class SpreadFakePlayersCommand {
+
+    public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        dispatcher.register(
+            Commands.literal("spreadfakeplayers")
+            .requires(source -> source.hasPermission(2))
+            .then(
+                Commands.argument("center", Vec2Argument.vec2())
+                .then(
+                    Commands.argument("spreadDistance", FloatArgumentType.floatArg(0.0F))
+                    .then(
+                        Commands.argument("maxRange", FloatArgumentType.floatArg(1.0F))
+                        .then(
+                            Commands.argument("respectTeams", BoolArgumentType.bool())
+                            .then(
+                                Commands.argument("amount", IntegerArgumentType.integer(1, 500))
+                                .executes(
+                                    ctx -> createFakePlayers(
+                                            ctx,
+                                            ctx.getSource().getLevel().getMaxBuildHeight(),
+                                            IntegerArgumentType.getInteger(ctx, "amount"),
+                                            ctx.getSource().getLevel()
+                                        )
+                                    )
+                                )
+                                .then(
+                                    Commands.argument("dimension", DimensionArgument.dimension())
+                                    .executes(
+                                        ctx -> createFakePlayers(
+                                                ctx,
+                                                ctx.getSource().getLevel().getMaxBuildHeight(),
+                                                IntegerArgumentType.getInteger(ctx, "amount"),
+                                                DimensionArgument.getDimension(ctx, "dimension")
+                                            )
+                                        )
+                                    )
+                                )
+                            )
+                        )
+                        .then(
+                            Commands.literal("under")
+                            .then(
+                                Commands.argument("maxHeight", IntegerArgumentType.integer())
+                                .then(
+                                    Commands.argument("respectTeams", BoolArgumentType.bool())
+                                    .then(
+                                        Commands.argument("amount", IntegerArgumentType.integer(1, 500))
+                                        .executes(
+                                            ctx -> createFakePlayers(
+                                                ctx,
+                                                IntegerArgumentType.getInteger(ctx, "maxHeight"),
+                                                IntegerArgumentType.getInteger(ctx, "amount"),
+                                                ctx.getSource().getLevel()
+                                            )
+                                        )
+                                        .then(
+                                            Commands.argument("dimension", DimensionArgument.dimension())
+                                            .executes(
+                                                ctx -> createFakePlayers(
+                                                    ctx,
+                                                    IntegerArgumentType.getInteger(ctx, "maxHeight"),
+                                                    IntegerArgumentType.getInteger(ctx, "amount"),
+                                                    DimensionArgument.getDimension(ctx, "dimension")
+                                                )
+                                            )
+                                        )
+                                    )
+                                )
+                            )
+                        )
+                    )
+                );
+    }
+
+    private static int createFakePlayers(
+        final CommandContext<CommandSourceStack> ctx,
+        final int maxHeight,
+        final int amount,
+        final ServerLevel dimension
+    ) throws CommandSyntaxException {
+        FakePlayerFactory.unloadLevel(dimension);
+
+        final ReferenceArraySet<ServerPlayer> fakePlayers = new ReferenceArraySet<>();
+        for (int i = 0; i < amount; i++) {
+            final var gameProfile = new GameProfile(UUID.randomUUID(), "FakePlayer" + String.format("%04d", i));
+            final var cookie = CommonListenerCookie.createInitial(gameProfile, false);
+            final var connection = new Connection(PacketFlow.SERVERBOUND);
+            new EmbeddedChannel(connection);
+            ServerPlayer fakePlayer = FakePlayerFactory.get(dimension, gameProfile);
+            fakePlayers.add(fakePlayer);
+
+            ctx.getSource().getServer().getPlayerList().placeNewPlayer(connection, fakePlayer, cookie, Optional.empty());
+        }
+
+        return spreadPlayers(
+            ctx.getSource(),
+            Vec2Argument.getVec2(ctx, "center"),
+            FloatArgumentType.getFloat(ctx, "spreadDistance"),
+            FloatArgumentType.getFloat(ctx, "maxRange"),
+            ctx.getSource().getLevel().getMaxBuildHeight(),
+            BoolArgumentType.getBool(ctx, "respectTeams"),
+            fakePlayers
+        );
+    }
+}
